<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analysis and Visualization of Gene Splice Variants and Transcriptome Data • splicejam</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Analysis and Visualization of Gene Splice Variants and Transcriptome Data">
<meta property="og:description" content="Jam Bioinformatics methods for RNA-seq and other genomics
    analysis methods. Sashimi coverage plots for RNA-seq, transcript isoform
    analysis methods to define detected isoforms from RNA-seq data, supporting
    methods for manipulating gene expression data.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">splicejam</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.53.900</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/basic-rnaseq-workflow.html">Splicejam basic RNAseq workflow</a>
    </li>
    <li>
      <a href="articles/create-a-sashimi-plot.html">Create a Sashimi Plot</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jmw86069/splicejam">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    

    
    
<!-- README.md is generated from README.Rmd. Please edit that file -->



<div id="splicejam" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#splicejam" class="anchor"></a>splicejam <img src="reference/figures/splicejam_logo.png" width="133px" height="154px" align="right" style="padding-left:10px;background-color:white;">
</h1></div>
<p>Splicejam was created to analyze and visualize RNA-seq and transcript isoform splicing data. Splicejam aims to provide sashimi plots with enough customizations to support publication-quality figures.</p>
<p>An example sashimi plot is shown below for the gene <em>Gria1</em>. Each panel shows transcript expression in a region of mouse hippocampus, where “CA1_CB” shows data that originated from cell bodies of CA1, and “CA2_CB” shows corresponding data from cell bodies of CA2.</p>
<p><br></p>
<p><img src="reference/figures/README-gria1_sashimi-1.png" width="1344"></p>
<figure><img src="reference/figures/mouse_hippocampus.png" width="25%" height="25%" align="right" style="padding-left:10px;background-color:white;"><figcaption>
A depiction of the CA1 and CA2 regions of mouse hippocampus is shown in Fig 4a from <a href="https://dx.doi.org/10.1038/nrn.2015.22">Dudek et al. Nature Reviews Neuroscience(2016)</a>
</figcaption></figure><p><br></p>
<div id="features-of-a-splicejam-sashimi-plot" class="section level2">
<h2 class="hasAnchor">
<a href="#features-of-a-splicejam-sashimi-plot" class="anchor"></a>Features of a splicejam sashimi plot:</h2>
<ul>
<li>The dark color polygons represent RNA transcript sequence coverage for exons in the Gria1 gene.</li>
<li>Wide ribbons are drawn to indicate splice junctions, areas where the sequences are aligned with a wide gap, typically spanning two exons.</li>
<li>The thickness of the ribbon (in height) represents the number of aligned reads that span the gap, usually about 70-80% the height of the adjacent exons.</li>
<li>The ribbons are shaded light to dark based upon how predominant the splice junction is from exon to exon. The darkest ribbons represent the most predominant path from exon to exon through the gene.</li>
<li>Introns are drawn with a fixed width (along the x-axis) and do not use genome coordinates. Introns are normally about 100 times wider than exons, and when drawn to scale obscure the detailed coverage of the exons. The x-axis labels indicate genome coordinates.</li>
</ul>
</div>
<div id="sashimi-plot-at-closer-range" class="section level2">
<h2 class="hasAnchor">
<a href="#sashimi-plot-at-closer-range" class="anchor"></a>Sashimi plot at closer range</h2>
<p>The x-axis range is adjusted in the following plot to show more detail around the differentially spliced exons. Notice the relative heights of the exons differ across CA1 and CA2, consistent with the corresponding changes in splice junctions. Clearly CA1 and CA2 favor different and mutually exclusive exons for Gria1.</p>
<p><img src="reference/figures/README-gria1_sashimi_zoom-1.png" width="768"></p>
<p>Incidentally, these two isoforms of Gria1 represent the <a href="https://www.sciencedirect.com/topics/neuroscience/gria1">“flop” and “flip”</a> forms of the AMPA receptor complex. In human, the genes Gria2, Gria3, and Gria4 comprise the AMPA receptor complex, and each genes has two isoforms with “flop” and “flip” designations.</p>
</div>
<div id="package-reference" class="section level2">
<h2 class="hasAnchor">
<a href="#package-reference" class="anchor"></a>Package Reference</h2>
<p>A full online function reference is available via the pkgdown documentation:</p>
<p>Full splicejam command reference: <a href="https://jmw86069.github.io/splicejam"></a><a href="https://jmw86069.github.io/splicejam">https://jmw86069.github.io/splicejam</a></p>
<p>The splicejam package is part of a suite of R packages called “jampack” which is available through GitHub, see <a href="https://github.com/jmw86069/jampack"></a><a href="https://github.com/jmw86069/jampack">https://github.com/jmw86069/jampack</a></p>
</div>
<div id="how-to-install" class="section level2">
<h2 class="hasAnchor">
<a href="#how-to-install" class="anchor"></a>How to install</h2>
<p>Install using the R package <code>devtools</code> and this command:</p>
<pre><code><a href="https://www.rdocumentation.org/packages/devtools/topics/reexports">devtools::install_github("jmw86069/splicejam")</a></code></pre>
</div>
<div id="about-sashimi-plots" class="section level2">
<h2 class="hasAnchor">
<a href="#about-sashimi-plots" class="anchor"></a>About sashimi plots</h2>
<p>Sashimi plots were originally envisioned by MISO, which introduced the innovative idea to compress intron width in order to reveal the detailed exon sequence coverage, and to highlight changes in splice junctions.</p>
<p>Katz, Y, Wang ET, Silterra J, Schwartz S, Wong B, Thorvaldsdóttir H, Robinson JT, Mesirov JP, Airoldi EM, Burge, CB. <em>Sashimi plots: Quantitative visualization of alternative isoform expression from RNA-seq data.</em> <a href="http://biorxiv.org/content/early/2014/02/11/002576">http://biorxiv.org/content/early/2014/02/11/002576</a></p>
<p>Splicejam re-envisions classic sashimi plots by displaying splice junctions as wide arcs proportional in size to the number of junction reads, making them directly comparable to exon coverages on the same plot. Inspired by:</p>
<ul>
<li>
<a href="https://en.wikipedia.org/wiki/Sankey_diagram">Sankey plots</a>,</li>
<li>
<a href="https://en.wikipedia.org/wiki/Flow_diagram">flow diagrams</a>,</li>
<li>
<a href="https://en.wikipedia.org/wiki/Alluvial_diagram">Alluvial diagrams</a>.</li>
</ul>
<p>These ribbons were ultimately made feasible to implement in R thanks to the package <a href="https://ggforce.data-imaginist.com/">ggforce</a> which implemented <a href="https://ggforce.data-imaginist.com/reference/geom_diagonal_wide.html">geom_diagonal_wide</a>.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/jmw86069/splicejam">https://​github.com/​jmw86069/​splicejam</a>
</li>
<li>Report a bug at <br><a href="http://github.com/jmw86069/splicejam/issues">http://​github.com/​jmw86069/​splicejam/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-2">GPL-2</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>James M. Ward <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-9510-2848" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" alt="ORCID" height="16"></a> </li>
<li>Shannon Farris <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-4473-1684" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" alt="ORCID" height="16"></a> </li>
</ul>
</div>

      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by James M. Ward, Shannon Farris.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
