<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Splicejam basic RNAseq workflow • splicejam</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Splicejam basic RNAseq workflow">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">splicejam</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.14.900</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/basic-rnaseq-workflow.html">Splicejam basic RNAseq workflow</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jmw86069/splicejam">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Splicejam basic RNAseq workflow</h1>
                        <h4 class="author">James M. Ward</h4>
            
            <h4 class="date">2019-02-06</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/jmw86069/splicejam/blob/master/vignettes/basic-rnaseq-workflow.Rmd"><code>vignettes/basic-rnaseq-workflow.Rmd</code></a></small>
      <div class="hidden name"><code>basic-rnaseq-workflow.Rmd</code></div>

    </div>

    
    
<p>This vignette is intended to demonstrate the basic analysis workflow for RNA-seq data.</p>
<div id="r-setup" class="section level2">
<h2 class="hasAnchor">
<a href="#r-setup" class="anchor"></a>R setup</h2>
<p>First, set up the R environment with basic packages loaded.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/message">suppressPackageStartupMessages</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(splicejam));
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/message">suppressPackageStartupMessages</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(colorjam));
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/message">suppressPackageStartupMessages</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(jamba));
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/message">suppressPackageStartupMessages</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(jamma));</code></pre></div>
</div>
<div id="salmon-quant-sf-gz-files" class="section level2">
<h2 class="hasAnchor">
<a href="#salmon-quant-sf-gz-files" class="anchor"></a>Salmon “quant.sf.gz” files</h2>
<p>This example begins with Salmon quantification files from the <code>tximportData</code> package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/message">suppressPackageStartupMessages</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(tximportData));
dir &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="dt">package=</span><span class="st">"tximportData"</span>);

<span class="co"># read sample annotations as a data.frame</span>
samples &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/file.path">file.path</a></span>(dir, <span class="st">"samples.txt"</span>), <span class="dt">header=</span><span class="ot">TRUE</span>);
samples;
<span class="co">#&gt;   pop center                assay    sample experiment       run</span>
<span class="co">#&gt; 1 TSI  UNIGE NA20503.1.M_111124_5 ERS185497  ERX163094 ERR188297</span>
<span class="co">#&gt; 2 TSI  UNIGE NA20504.1.M_111124_7 ERS185242  ERX162972 ERR188088</span>
<span class="co">#&gt; 3 TSI  UNIGE NA20505.1.M_111124_6 ERS185048  ERX163009 ERR188329</span>
<span class="co">#&gt; 4 TSI  UNIGE NA20507.1.M_111124_7 ERS185412  ERX163158 ERR188288</span>
<span class="co">#&gt; 5 TSI  UNIGE NA20508.1.M_111124_2 ERS185362  ERX163159 ERR188021</span>
<span class="co">#&gt; 6 TSI  UNIGE NA20514.1.M_111124_4 ERS185217  ERX163062 ERR188356</span>

<span class="co"># define Salmon import files</span>
files &lt;-<span class="st"> </span>jamba<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/jamba/topics/nameVector">nameVector</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/file.path">file.path</a></span>(dir, <span class="st">"salmon"</span>, samples<span class="op">$</span>run, <span class="st">"quant.sf.gz"</span>),
   samples<span class="op">$</span>run);
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>(files);
<span class="co">#&gt;                                                                                           files</span>
<span class="co">#&gt; ERR188297 /Users/wardjm/Library/R/3.5/library/tximportData/extdata/salmon/ERR188297/quant.sf.gz</span>
<span class="co">#&gt; ERR188088 /Users/wardjm/Library/R/3.5/library/tximportData/extdata/salmon/ERR188088/quant.sf.gz</span>
<span class="co">#&gt; ERR188329 /Users/wardjm/Library/R/3.5/library/tximportData/extdata/salmon/ERR188329/quant.sf.gz</span>
<span class="co">#&gt; ERR188288 /Users/wardjm/Library/R/3.5/library/tximportData/extdata/salmon/ERR188288/quant.sf.gz</span>
<span class="co">#&gt; ERR188021 /Users/wardjm/Library/R/3.5/library/tximportData/extdata/salmon/ERR188021/quant.sf.gz</span>
<span class="co">#&gt; ERR188356 /Users/wardjm/Library/R/3.5/library/tximportData/extdata/salmon/ERR188356/quant.sf.gz</span></code></pre></div>
</div>
<div id="transcript-to-gene-from-gtf" class="section level2">
<h2 class="hasAnchor">
<a href="#transcript-to-gene-from-gtf" class="anchor"></a>Transcript-to-gene from GTF</h2>
<p>The transcript-to-gene association will be stored in a data.frame <code>tx2geneDF</code>, in this case using <code><a href="../reference/makeTx2geneFromGtf.html">splicejam::makeTx2geneFromGtf()</a></code>. For other ways of creating a <code>tx2gene</code> data.frame, see the vignette from the <code><a href="https://www.rdocumentation.org/packages/tximport/topics/tximport">tximport::tximport</a></code> package.</p>
<p>For the purpose of this example, the GTF file is used from the <code>tximportData</code> package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gtfFile &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(
   <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">list.files</a></span>(<span class="dt">pattern=</span><span class="st">"gtf.gz"</span>,
      <span class="dt">path=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata/salmon_dm"</span>, <span class="dt">package=</span><span class="st">"tximportData"</span>),
      <span class="dt">full.names=</span><span class="ot">TRUE</span>),
   <span class="dv">1</span>);
gtfFile;
<span class="co">#&gt; [1] "/Users/wardjm/Library/R/3.5/library/tximportData/extdata/salmon_dm/Drosophila_melanogaster.BDGP6.92.gtf.gz"</span>
tx2geneDF &lt;-<span class="st"> </span><span class="kw"><a href="../reference/makeTx2geneFromGtf.html">makeTx2geneFromGtf</a></span>(<span class="dt">GTF=</span>gtfFile);
<span class="co">#&gt; Warning: package 'data.table' was built under R version 3.5.2</span>
<span class="co">#&gt; Warning in data.table::fread(GTF, sep = "\t", autostart = 20, nrows =</span>
<span class="co">#&gt; nrows, : 'autostart' is now deprecated and ignored. Consider skip='string'</span>
<span class="co">#&gt; or skip=n</span>
<span class="co">#&gt; Warning in data.table::fread(GTF, sep = "\t", autostart = 20, nrows =</span>
<span class="co">#&gt; nrows, : Detected 1 column names but the data has 9 columns (i.e. invalid</span>
<span class="co">#&gt; file). Added 8 extra default column names at the end.</span>
txColname &lt;-<span class="st"> "transcript_id"</span>;
geneColname &lt;-<span class="st"> "gene_name"</span>;
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(tx2geneDF));
<span class="co">#&gt;       gene_id      gene_name transcript_id</span>
<span class="co">#&gt; 1 FBgn0000003 7SLRNA:CR32864   FBtr0081624</span>
<span class="co">#&gt; 2 FBgn0000008              a   FBtr0100521</span>
<span class="co">#&gt; 3 FBgn0000008              a   FBtr0071763</span>
<span class="co">#&gt; 4 FBgn0000008              a   FBtr0342981</span>
<span class="co">#&gt; 5 FBgn0000008              a   FBtr0071764</span>
<span class="co">#&gt; 6 FBgn0000014          abd-A   FBtr0306337</span>

## Alternatively, we load a prepared file from tximportData
tx2geneDF &lt;-<span class="st"> </span>data.table<span class="op">::</span><span class="kw">fread</span>(
   <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/file.path">file.path</a></span>(dir, <span class="st">"tx2gene.gencode.v27.csv"</span>),
   <span class="dt">sep=</span><span class="st">","</span>,
   <span class="dt">data.table=</span><span class="ot">FALSE</span>);
txColname &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(tx2geneDF)[<span class="dv">1</span>];
geneColname &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(tx2geneDF)[<span class="dv">2</span>];</code></pre></div>
</div>
<div id="import-salmon-transcript-data" class="section level2">
<h2 class="hasAnchor">
<a href="#import-salmon-transcript-data" class="anchor"></a>Import Salmon transcript data</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">txiTx &lt;-<span class="st"> </span>tximport<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/tximport/topics/tximport">tximport</a></span>(files,
   <span class="dt">type=</span><span class="st">"salmon"</span>,
   <span class="dt">txOut=</span><span class="ot">TRUE</span>);
<span class="co">#&gt; reading in files with read_tsv</span>
<span class="co">#&gt; 1 2 3 4 5 6</span>

## Check the data returned
jamba<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/jamba/topics/sdim">sdim</a></span>(txiTx);
<span class="co">#&gt;                       rows cols     class</span>
<span class="co">#&gt; abundance           200401    6    matrix</span>
<span class="co">#&gt; counts              200401    6    matrix</span>
<span class="co">#&gt; length              200401    6    matrix</span>
<span class="co">#&gt; countsFromAbundance      1      character</span></code></pre></div>
<div id="summarizedexperiment-for-transcript-data" class="section level3">
<h3 class="hasAnchor">
<a href="#summarizedexperiment-for-transcript-data" class="anchor"></a>SummarizedExperiment for transcript data</h3>
<p>It is helpful and consistent with other Bioconductor workflows to use a commonly-used object “SummarizedExperiment”. This object stores gene data, biological sample data, and one or more data matrices in one convenient format.</p>
<blockquote>
<p>Note: During this step, data is log2-transformed using the format <code><a href="https://www.rdocumentation.org/packages/base/topics/Log">log2(1+x)</a></code>.</p>
</blockquote>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Create SummarizedExperiment transcript object
<span class="cf">if</span> (<span class="op">!</span><span class="st">"package:SummarizedExperiment"</span> <span class="op">%in%</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/search">search</a></span>()) {
   <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/message">suppressPackageStartupMessages</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(SummarizedExperiment));
}
<span class="co">#&gt; Warning: package 'BiocParallel' was built under R version 3.5.2</span>

assayNames &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sets">intersect</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"abundance"</span>,<span class="st">"counts"</span>), <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(txiTx));
geneMatch &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/match">match</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(txiTx[[assayNames[<span class="dv">1</span>]]]),
   tx2geneDF[,txColname]);
sampleDF &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>(<span class="dt">Sample=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(txiTx[[assayNames[<span class="dv">1</span>]]]),
   <span class="dt">Group=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"A"</span>, <span class="st">"B"</span>), <span class="dt">length.out=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">ncol</a></span>(txiTx[[assayNames[<span class="dv">1</span>]]])));
TxSE &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/RangedSummarizedExperiment-class">SummarizedExperiment</a></span>(
   <span class="dt">assays=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">lapply</a></span>(txiTx[assayNames], <span class="cf">function</span>(x){<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log2</a></span>(<span class="dv">1</span><span class="op">+</span>x)}),
   <span class="dt">rowData=</span><span class="kw">DataFrame</span>(tx2geneDF[geneMatch,,<span class="dt">drop=</span><span class="ot">FALSE</span>]),
   <span class="dt">colData=</span><span class="kw">DataFrame</span>(sampleDF)
)</code></pre></div>
</div>
</div>
<div id="import-salmon-gene-data-using-tx2genedf" class="section level2">
<h2 class="hasAnchor">
<a href="#import-salmon-gene-data-using-tx2genedf" class="anchor"></a>Import Salmon gene data using tx2geneDF</h2>
<p>If data was imported at the transcript level (above) then it can be summarized at the gene level without re-importing from the source files. Alternatively, it can be imported from source files without importing at the transcript level. Both methods are shown below.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">##
<span class="cf">if</span> (<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/exists">exists</a></span>(<span class="st">"txiTx"</span>)) {
   txiGene &lt;-<span class="st"> </span>tximport<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/tximport/topics/tximport">summarizeToGene</a></span>(txiTx,
      <span class="dt">tx2gene=</span>tx2geneDF[,<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(txColname,geneColname)]);
} <span class="cf">else</span> {
   txiGene &lt;-<span class="st"> </span>tximport<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/tximport/topics/tximport">tximport</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unname">unname</a></span>(files),
      <span class="dt">type=</span><span class="st">"salmon"</span>,
      <span class="dt">tx2gene=</span>tx2geneDF[,<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(txColname,geneColname)]);
}
<span class="co">#&gt; summarizing abundance</span>
<span class="co">#&gt; summarizing counts</span>
<span class="co">#&gt; summarizing length</span>

## Check the data returned
jamba<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/jamba/topics/sdim">sdim</a></span>(txiGene);
<span class="co">#&gt;                      rows cols     class</span>
<span class="co">#&gt; abundance           58288    6    matrix</span>
<span class="co">#&gt; counts              58288    6    matrix</span>
<span class="co">#&gt; length              58288    6    matrix</span>
<span class="co">#&gt; countsFromAbundance     1      character</span></code></pre></div>
<div id="summarizedexperiment-for-gene-data" class="section level3">
<h3 class="hasAnchor">
<a href="#summarizedexperiment-for-gene-data" class="anchor"></a>SummarizedExperiment for gene data</h3>
<p>It is helpful and consistent with other Bioconductor workflows to use a commonly-used object “SummarizedExperiment”. This object stores gene data, biological sample data, and one or more data matrices in one convenient format.</p>
<blockquote>
<p>Note: This step is a convenient time to annotate the gene data.frame, for example the steps below will use gene-related <code><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames(tx2geneDF)</a></code> where possible. If a “gene_name” or “gene_symbol” colname does not exist, this step would be a good time to add that data.</p>
</blockquote>
<blockquote>
<p>Note: During this step, data is log2-transformed using the format <code><a href="https://www.rdocumentation.org/packages/base/topics/Log">log2(1+x)</a></code>.</p>
</blockquote>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Create SummarizedExperiment transcript object

assayNamesG &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sets">intersect</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"abundance"</span>,<span class="st">"counts"</span>), <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(txiGene));
geneMatchG &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/match">match</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(txiGene[[assayNames[<span class="dv">1</span>]]]),
   tx2geneDF[,geneColname]);
geneColnames &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unique">unique</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(geneColname,
   jamba<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/jamba/topics/unvigrep">unvigrep</a></span>(<span class="st">"^trans|^tx"</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(tx2geneDF))));
sampleDF &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>(<span class="dt">Sample=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(txiGene[[assayNames[<span class="dv">1</span>]]]),
   <span class="dt">Group=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"A"</span>, <span class="st">"B"</span>), <span class="dt">length.out=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">ncol</a></span>(txiGene[[assayNames[<span class="dv">1</span>]]])));
GeneSE &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/RangedSummarizedExperiment-class">SummarizedExperiment</a></span>(
   <span class="dt">assays=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">lapply</a></span>(txiGene[assayNamesG], <span class="cf">function</span>(x){<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log2</a></span>(<span class="dv">1</span><span class="op">+</span>x)}),
   <span class="dt">rowData=</span><span class="kw">DataFrame</span>(tx2geneDF[geneMatchG,geneColnames,<span class="dt">drop=</span><span class="ot">FALSE</span>]),
   <span class="dt">colData=</span><span class="kw">DataFrame</span>(sampleDF)
)</code></pre></div>
</div>
</div>
<div id="transcript-level-analysis-steps" class="section level2">
<h2 class="hasAnchor">
<a href="#transcript-level-analysis-steps" class="anchor"></a>Transcript-level analysis steps</h2>
<div id="define-detected-transcripts" class="section level3">
<h3 class="hasAnchor">
<a href="#define-detected-transcripts" class="anchor"></a>Define detected transcripts</h3>
<p>Often the RNA-seq data includes a number of transcripts for which there is no detectable signal. Several filtering methods are encapsulated into a function <code><a href="../reference/defineDetectedTx.html">defineDetectedTx()</a></code> to filter transcript isoform data which should be considered below the effective limit of detection.</p>
<p>At this step, group mean values are used for filtering, however it can be performed without grouping, by setting <code>groups=colnames(TxSE)</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cutoffTxExpr &lt;-<span class="st"> </span><span class="dv">7</span>;
cutoffTxTPMExpr &lt;-<span class="st"> </span><span class="dv">1</span>;
cutoffTxPctMax &lt;-<span class="st"> </span><span class="dv">10</span>;
detectedTxTPML &lt;-<span class="st"> </span><span class="kw"><a href="../reference/defineDetectedTx.html">defineDetectedTx</a></span>(
   <span class="dt">iMatrixTx=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class">assays</a></span>(TxSE)[[<span class="st">"counts"</span>]],
   <span class="dt">iMatrixTxTPM=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class">assays</a></span>(TxSE)[[<span class="st">"abundance"</span>]],
   <span class="dt">groups=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/jamba/topics/nameVector">nameVector</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class">colData</a></span>(TxSE)<span class="op">$</span>Group, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(TxSE)),
   <span class="dt">cutoffTxPctMax=</span>cutoffTxPctMax,
   <span class="dt">cutoffTxExpr=</span>cutoffTxExpr,
   <span class="dt">cutoffTxTPMExpr=</span>cutoffTxTPMExpr,
   <span class="dt">tx2geneDF=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/jamba/topics/renameColumn">renameColumn</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class">rowData</a></span>(TxSE),
     <span class="dt">from=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(geneColname,txColname),
     <span class="dt">to=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"gene_name"</span>,<span class="st">"transcript_id"</span>)),
   <span class="dt">useMedian=</span><span class="ot">FALSE</span>,
   <span class="dt">verbose=</span><span class="ot">FALSE</span>);
detectedTx &lt;-<span class="st"> </span>detectedTxTPML<span class="op">$</span>detectedTx;
numDetectedTx &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(detectedTx);
detectedGenes &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/jamba/topics/mixedSort">mixedSort</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unique">unique</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class">rowData</a></span>(TxSE[detectedTx,])[,geneColname]));
numDetectedGenes &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(detectedGenes);
<span class="kw"><a href="https://www.rdocumentation.org/packages/jamba/topics/printDebug">printDebug</a></span>(<span class="st">"Defined "</span>,
   <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/format">format</a></span>(<span class="dt">big.mark=</span><span class="st">","</span>, numDetectedTx),
   <span class="st">" detected Tx representing "</span>,
   <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/format">format</a></span>(<span class="dt">big.mark=</span><span class="st">","</span>, numDetectedGenes),
   <span class="st">" genes."</span>);
<span class="co">#&gt; Warning: package 'colorspace' was built under R version 3.5.2</span>
<span class="co">#&gt; ##  (17:14:13) 06Feb2019:  Defined 39,111 detected Tx representing 15,168 genes.</span></code></pre></div>
<ul>
<li>**{r format(numDetectedTx, scientific=FALSE, big.mark=“,”)} detected transcripts</li>
<li>**{r format(numDetectedGenes, scientific=FALSE, big.mark=“,”)} detected genes</li>
</ul>
<div id="determining-reasonable-cutoff-values" class="section level4">
<h4 class="hasAnchor">
<a href="#determining-reasonable-cutoff-values" class="anchor"></a>Determining reasonable cutoff values</h4>
<p>Some of the cutoff values can be visualized by plotting the TPM versus the pseudocounts, shown below.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## For this plot, filter out all rows where TPM and abundance are zero.
iWhich &lt;-<span class="st"> </span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class">assays</a></span>(TxSE[,<span class="dv">1</span>])[[<span class="st">"counts"</span>]] <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span> <span class="op">&amp;</span>
<span class="st">      </span><span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class">assays</a></span>(TxSE[,<span class="dv">1</span>])[[<span class="st">"abundance"</span>]] <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>);
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/par">par</a></span>(<span class="st">"mfrow"</span>=<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>,<span class="dv">1</span>), <span class="st">"mar"</span>=<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">5</span>,<span class="dv">4</span>,<span class="dv">4</span>,<span class="dv">2</span>), <span class="st">"oma"</span>=<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>));
<span class="kw"><a href="https://www.rdocumentation.org/packages/jamba/topics/plotSmoothScatter">plotSmoothScatter</a></span>(<span class="dt">x=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log2</a></span>(<span class="dv">1</span><span class="op">+</span><span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class">assays</a></span>(TxSE[iWhich,<span class="dv">1</span>])[[<span class="st">"counts"</span>]]),
   <span class="dt">y=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log2</a></span>(<span class="dv">1</span><span class="op">+</span><span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class">assays</a></span>(TxSE[iWhich,<span class="dv">1</span>])[[<span class="st">"abundance"</span>]]),
   <span class="dt">xlab=</span><span class="st">"counts"</span>,
   <span class="dt">ylab=</span><span class="st">"abundance"</span>,
   <span class="dt">xaxt=</span><span class="st">"n"</span>,
   <span class="dt">yaxt=</span><span class="st">"n"</span>);
<span class="co">#&gt; Warning in NSBS(i, x, exact = exact, strict.upper.bound = !allow.append, :</span>
<span class="co">#&gt; subscript is an array, passing it thru as.vector() first</span>

<span class="co">#&gt; Warning in NSBS(i, x, exact = exact, strict.upper.bound = !allow.append, :</span>
<span class="co">#&gt; subscript is an array, passing it thru as.vector() first</span>
jamba<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/jamba/topics/minorLogTicksAxis">minorLogTicksAxis</a></span>(<span class="dv">1</span>,
   <span class="dt">logBase=</span><span class="dv">2</span>, <span class="dt">displayBase=</span><span class="dv">10</span>, <span class="dt">offset=</span><span class="dv">1</span>);
<span class="co">#&gt; ##  (17:14:13) 06Feb2019:  adjusting logBase to displayBase.</span>
<span class="co">#&gt; Warning in any(majorLabels): coercing argument of type 'double' to logical</span>
<span class="co">#&gt; ##  (17:14:13) 06Feb2019:  majorTicks:-3.459432,-1.000000,0.000000,1.000000,3.459432,6.658211 </span>
<span class="co">#&gt; ##  (17:14:13) 06Feb2019:  majorLabels:-10,-1,0,1,10,100</span>
jamba<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/jamba/topics/minorLogTicksAxis">minorLogTicksAxis</a></span>(<span class="dv">2</span>,
   <span class="dt">logBase=</span><span class="dv">2</span>, <span class="dt">displayBase=</span><span class="dv">10</span>, <span class="dt">offset=</span><span class="dv">1</span>);
<span class="co">#&gt; ##  (17:14:13) 06Feb2019:  adjusting logBase to displayBase.</span>
<span class="co">#&gt; Warning in any(majorLabels): coercing argument of type 'double' to logical</span>
<span class="co">#&gt; ##  (17:14:14) 06Feb2019:  majorTicks:-3.459432,-1.000000,0.000000,1.000000,3.459432,6.658211 </span>
<span class="co">#&gt; ##  (17:14:14) 06Feb2019:  majorLabels:-10,-1,0,1,10,100</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/abline">abline</a></span>(<span class="dt">v=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log2</a></span>(<span class="dv">1</span><span class="op">+</span>cutoffTxExpr),
   <span class="dt">h=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log2</a></span>(<span class="dv">1</span><span class="op">+</span>cutoffTxTPMExpr),
   <span class="dt">lty=</span><span class="st">"dashed"</span>,
   <span class="dt">col=</span><span class="st">"red"</span>);</code></pre></div>
<div class="figure">
<img src="basic-rnaseq-workflow_files/figure-html/tpm_counts-1.png" alt="Salmon pseudocounts versus TPM" width="672"><p class="caption">
Salmon pseudocounts versus TPM
</p>
</div>
<p>Another method to determine suitable cutoff values is to create an MA-plot of the data, and look for the expression below which the variability becomes very high.</p>
</div>
</div>
<div id="optional-data-normalization" class="section level3">
<h3 class="hasAnchor">
<a href="#optional-data-normalization" class="anchor"></a>Optional data normalization</h3>
<p>Typically, RNA-seq data abundances are affected by non-experimental factors such as total mapped reads, sample quality, sequence library efficiency, etc. Despite the fact that metrics such as FPKM and TPM are intended to reduce or remove the variability between samples, there are sometimes unavoidable differences which are apparently not biological in origin.</p>
<p>A full description of data normalization is beyond the scope of this document, however the recommended approach to review data is to use the <code><a href="https://www.rdocumentation.org/packages/jamma/topics/jammaplot">jamma::jammaplot()</a></code> MA-plot functions, as shown above.</p>
<div id="ma-plots-of-pseudocount-data" class="section level4">
<h4 class="hasAnchor">
<a href="#ma-plots-of-pseudocount-data" class="anchor"></a>MA-plots of pseudocount data</h4>
<p>The <code>jammaplot</code> R package is available on Github:</p>
<blockquote>
<p><code><a href="https://www.rdocumentation.org/packages/devtools/topics/reexports">devtools::install_github("jmw86069/jamma")</a></code></p>
</blockquote>
<p>First show the pseudocount data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="cf">if</span> (<span class="op">!</span><span class="st">"package:jamma"</span> <span class="op">%in%</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/search">search</a></span>()) {
   <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/message">suppressPackageStartupMessages</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(jamma));
}
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/par">par</a></span>(<span class="st">"oma"</span>=<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">3</span>,<span class="dv">0</span>));
jamma<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/jamma/topics/jammaplot">jammaplot</a></span>(
   <span class="kw"><a href="https://www.rdocumentation.org/packages/jamba/topics/noiseFloor">noiseFloor</a></span>(<span class="dt">minimum=</span><span class="fl">0.0001</span>,
      <span class="dt">newValue=</span><span class="ot">NA</span>,
      <span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class">assays</a></span>(TxSE)[[<span class="st">"counts"</span>]]),
   <span class="dt">transFactor=</span><span class="fl">0.3</span>,
   <span class="dt">ylim=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="op">-</span><span class="dv">3</span>,<span class="dv">3</span>),
   <span class="dt">ablineV=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log2</a></span>(<span class="dv">1</span><span class="op">+</span>cutoffTxExpr),
   <span class="dt">titleBoxColor=</span>colorjam<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/colorjam/topics/group2colors">group2colors</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class">colData</a></span>(TxSE)<span class="op">$</span>Group),
   <span class="dt">groupSuffix=</span><span class="st">""</span>,
   <span class="dt">titleCexFactor=</span><span class="fl">1.2</span>,
   <span class="dt">useMean=</span><span class="ot">TRUE</span>);
<span class="co">#&gt; Warning in rep(x, length.out = n): 'x' is NULL so the result will be NULL</span>
<span class="co">#&gt; Warning in rep(y, length.out = n): 'x' is NULL so the result will be NULL</span>
<span class="co">#&gt; Warning in text.default(x = textX[whichLabels], y = textY[whichLabels], :</span>
<span class="co">#&gt; "outer" is not a graphical parameter</span>
<span class="co">#&gt; Warning in rep(x, length.out = n): 'x' is NULL so the result will be NULL</span>
<span class="co">#&gt; Warning in rep(y, length.out = n): 'x' is NULL so the result will be NULL</span>
<span class="co">#&gt; Warning in text.default(x = textX[whichLabels], y = textY[whichLabels], :</span>
<span class="co">#&gt; "outer" is not a graphical parameter</span>
<span class="co">#&gt; Warning in rep(x, length.out = n): 'x' is NULL so the result will be NULL</span>
<span class="co">#&gt; Warning in rep(y, length.out = n): 'x' is NULL so the result will be NULL</span>
<span class="co">#&gt; Warning in text.default(x = textX[whichLabels], y = textY[whichLabels], :</span>
<span class="co">#&gt; "outer" is not a graphical parameter</span>
<span class="co">#&gt; Warning in rep(x, length.out = n): 'x' is NULL so the result will be NULL</span>
<span class="co">#&gt; Warning in rep(y, length.out = n): 'x' is NULL so the result will be NULL</span>
<span class="co">#&gt; Warning in text.default(x = textX[whichLabels], y = textY[whichLabels], :</span>
<span class="co">#&gt; "outer" is not a graphical parameter</span>
<span class="co">#&gt; Warning in rep(x, length.out = n): 'x' is NULL so the result will be NULL</span>
<span class="co">#&gt; Warning in rep(y, length.out = n): 'x' is NULL so the result will be NULL</span>
<span class="co">#&gt; Warning in text.default(x = textX[whichLabels], y = textY[whichLabels], :</span>
<span class="co">#&gt; "outer" is not a graphical parameter</span>
<span class="co">#&gt; Warning in rep(x, length.out = n): 'x' is NULL so the result will be NULL</span>
<span class="co">#&gt; Warning in rep(y, length.out = n): 'x' is NULL so the result will be NULL</span>
<span class="co">#&gt; Warning in text.default(x = textX[whichLabels], y = textY[whichLabels], :</span>
<span class="co">#&gt; "outer" is not a graphical parameter</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/title">title</a></span>(<span class="dt">outer=</span><span class="ot">TRUE</span>,
   <span class="dt">main=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(<span class="st">"MA-plot using Salmon pseudocounts</span><span class="ch">\n</span><span class="st">"</span>,
      <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(<span class="st">"Showing pseudocount cutoff value approximately "</span>,
         <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log2</a></span>(<span class="dv">1</span><span class="op">+</span>cutoffTxExpr))));</code></pre></div>
<p><img src="basic-rnaseq-workflow_files/figure-html/jammaplot_counts-1.png" width="480"></p>
<p>In the MA-plots above, one expects each sample to have roughly horizontal distribution, centered at <code>y=0</code>, which means the median transcript expression is roughly unchanged across all samples(<code>y=0</code>), and that the signal is roughly consistent from the low to high end of expression (horizontal).</p>
<p>Whenever a sample has bulk of its expression above (or below) <code>y=0</code>, it suggests there is overall higher (or lower) signal for all transcripts in that sample, which is counter to most assumptions of data normalization.</p>
</div>
<div id="ma-plots-of-tpm-data" class="section level4">
<h4 class="hasAnchor">
<a href="#ma-plots-of-tpm-data" class="anchor"></a>MA-plots of TPM data</h4>
<p>Next show the TPM data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/par">par</a></span>(<span class="st">"oma"</span>=<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">3</span>,<span class="dv">0</span>));
jamma<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/jamma/topics/jammaplot">jammaplot</a></span>(
   <span class="kw"><a href="https://www.rdocumentation.org/packages/jamba/topics/noiseFloor">noiseFloor</a></span>(<span class="dt">minimum=</span><span class="fl">0.0001</span>,
      <span class="dt">newValue=</span><span class="ot">NA</span>,
      <span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class">assays</a></span>(TxSE)[[<span class="st">"abundance"</span>]]),
   <span class="dt">transFactor=</span><span class="fl">0.3</span>,
   <span class="dt">ylim=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="op">-</span><span class="dv">3</span>,<span class="dv">3</span>),
   <span class="dt">ablineV=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log2</a></span>(<span class="dv">1</span><span class="op">+</span>cutoffTxTPMExpr),
   <span class="dt">titleBoxColor=</span>colorjam<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/colorjam/topics/group2colors">group2colors</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class">colData</a></span>(TxSE)<span class="op">$</span>Group),
   <span class="dt">groupSuffix=</span><span class="st">""</span>,
   <span class="dt">displayMAD=</span><span class="ot">TRUE</span>,
   <span class="dt">titleCexFactor=</span><span class="fl">1.2</span>,
   <span class="dt">useMean=</span><span class="ot">TRUE</span>);
<span class="co">#&gt; Warning in rep(x, length.out = n): 'x' is NULL so the result will be NULL</span>
<span class="co">#&gt; Warning in rep(y, length.out = n): 'x' is NULL so the result will be NULL</span>
<span class="co">#&gt; Warning in text.default(x = textX[whichLabels], y = textY[whichLabels], :</span>
<span class="co">#&gt; "outer" is not a graphical parameter</span>
<span class="co">#&gt; Warning in rep(x, length.out = n): 'x' is NULL so the result will be NULL</span>
<span class="co">#&gt; Warning in rep(y, length.out = n): 'x' is NULL so the result will be NULL</span>
<span class="co">#&gt; Warning in text.default(x = textX[whichLabels], y = textY[whichLabels], :</span>
<span class="co">#&gt; "outer" is not a graphical parameter</span>
<span class="co">#&gt; Warning in rep(x, length.out = n): 'x' is NULL so the result will be NULL</span>
<span class="co">#&gt; Warning in rep(y, length.out = n): 'x' is NULL so the result will be NULL</span>
<span class="co">#&gt; Warning in text.default(x = textX[whichLabels], y = textY[whichLabels], :</span>
<span class="co">#&gt; "outer" is not a graphical parameter</span>
<span class="co">#&gt; Warning in rep(x, length.out = n): 'x' is NULL so the result will be NULL</span>
<span class="co">#&gt; Warning in rep(y, length.out = n): 'x' is NULL so the result will be NULL</span>
<span class="co">#&gt; Warning in text.default(x = textX[whichLabels], y = textY[whichLabels], :</span>
<span class="co">#&gt; "outer" is not a graphical parameter</span>
<span class="co">#&gt; Warning in rep(x, length.out = n): 'x' is NULL so the result will be NULL</span>
<span class="co">#&gt; Warning in rep(y, length.out = n): 'x' is NULL so the result will be NULL</span>
<span class="co">#&gt; Warning in text.default(x = textX[whichLabels], y = textY[whichLabels], :</span>
<span class="co">#&gt; "outer" is not a graphical parameter</span>
<span class="co">#&gt; Warning in rep(x, length.out = n): 'x' is NULL so the result will be NULL</span>
<span class="co">#&gt; Warning in rep(y, length.out = n): 'x' is NULL so the result will be NULL</span>
<span class="co">#&gt; Warning in text.default(x = textX[whichLabels], y = textY[whichLabels], :</span>
<span class="co">#&gt; "outer" is not a graphical parameter</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/title">title</a></span>(<span class="dt">outer=</span><span class="ot">TRUE</span>,
   <span class="dt">main=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(<span class="st">"MA-plot using Salmon TPM</span><span class="ch">\n</span><span class="st">"</span>,
      <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(<span class="st">"Showing TPM cutoff value approximately "</span>,
         <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log2</a></span>(<span class="dv">1</span><span class="op">+</span>cutoffTxTPMExpr))));</code></pre></div>
<p><img src="basic-rnaseq-workflow_files/figure-html/jammaplot_tpm-1.png" width="480"></p>
<p>The typical remedy for certain MA-plot patterns are summarized briefly below:</p>
<ol style="list-style-type: decimal">
<li>If expression is roughly horizontal for all samples, with all samples centered closely at <code>y=0</code>, then no normalization is required.</li>
<li>If expression is roughly horizontal for all samples, with some samples centered above or below <code>y=0</code>, then median normalization is recommended.</li>
<li>If expression is not horizontal for all samples, where some subset of samples show a slope upward, or downward from left-to-right on the MA-plot panel, the usual remedy is quantile normalization.</li>
<li>If the variability above/below <code>y=0</code> is substantially higher for some samples than the majority of all samples, these samples may be considered technical outliers. Use <code>displayMAD</code> on the <code><a href="https://www.rdocumentation.org/packages/jamma/topics/jammaplot">jammaplot()</a></code> to view the relative variability across samples. A MAD factor threshold of 2xMAD is typically enough to identify outliers among biological replicates, or threshold 5xMAD across all samples.</li>
</ol>
</div>
</div>
<div id="differential-isoform-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#differential-isoform-analysis" class="anchor"></a>Differential isoform analysis</h3>
<p>Differential isoform analysis is carried out using the function <code><a href="../reference/runDiffSplice.html">splicejam::runDiffSplice()</a></code>, which performs several convenient steps including the core function <code><a href="https://www.rdocumentation.org/packages/limma/topics/diffSplice">limma::diffSplice()</a></code>.</p>
<div id="design-and-contrast-matrices" class="section level4">
<h4 class="hasAnchor">
<a href="#design-and-contrast-matrices" class="anchor"></a>Design and Contrast matrices</h4>
<p>Before statistical comparisons can be performed, the design matrix and contrast matrix must be defined. The function <code><a href="../reference/groups2contrasts.html">groups2contrasts()</a></code> makes it convenient to convert group names to a set of contrasts. The group names are assumed to have underscore “_" as a delimiter between factors, for example <code>"Wildtype_Control"</code> and <code>"Mutant_Treated"</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Design matrix is defined by the sample groups
iSamples &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(TxSE);
iGroups &lt;-<span class="st"> </span>jamba<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/jamba/topics/nameVector">nameVector</a></span>(
   <span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class">colData</a></span>(TxSE[,iSamples])<span class="op">$</span>Group,
      iSamples);

## Use the function groups2contrasts()
iDC &lt;-<span class="st"> </span><span class="kw"><a href="../reference/groups2contrasts.html">groups2contrasts</a></span>(iGroups, <span class="dt">returnDesign=</span><span class="ot">TRUE</span>);
iDesign &lt;-<span class="st"> </span>iDC<span class="op">$</span>iDesign;
iContrasts &lt;-<span class="st"> </span>iDC<span class="op">$</span>iContrasts;

<span class="kw"><a href="https://www.rdocumentation.org/packages/jamba/topics/printDebug">printDebug</a></span>(<span class="st">"iDesign:"</span>);
<span class="co">#&gt; ##  (17:14:20) 06Feb2019:  iDesign:</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(iDesign);
<span class="co">#&gt;           A B</span>
<span class="co">#&gt; ERR188297 1 0</span>
<span class="co">#&gt; ERR188088 0 1</span>
<span class="co">#&gt; ERR188329 1 0</span>
<span class="co">#&gt; ERR188288 0 1</span>
<span class="co">#&gt; ERR188021 1 0</span>
<span class="co">#&gt; ERR188356 0 1</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/jamba/topics/printDebug">printDebug</a></span>(<span class="st">"iContrasts:"</span>);
<span class="co">#&gt; ##  (17:14:20) 06Feb2019:  iContrasts:</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(iContrasts);
<span class="co">#&gt;       Contrasts</span>
<span class="co">#&gt; Levels B-A</span>
<span class="co">#&gt;      A  -1</span>
<span class="co">#&gt;      B   1</span>

## Alternative method
<span class="cf">if</span> (<span class="dv">1</span> <span class="op">==</span><span class="st"> </span><span class="dv">2</span>) {
   iDesign &lt;-<span class="st"> </span>stats<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/model.matrix">model.matrix</a></span>(<span class="op">~</span><span class="dv">0</span><span class="op">+</span>iGroups);
   <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(iDesign) &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/levels">levels</a></span>(iGroups);
   <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(iDesign) &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(iGroups);
   <span class="co"># Alternative format, useful for custom contrasts</span>
   iContrasts &lt;-<span class="st"> </span>limma<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/limma/topics/makeContrasts">makeContrasts</a></span>(
      <span class="dt">contrasts=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(
         <span class="st">"CA2_DE-CA2_CB"</span>,
         <span class="st">"CA1_DE-CA1_CB"</span>),
      <span class="dt">levels=</span>iDesign);
}</code></pre></div>
<p>Note that it is important to keep the colnames(TxSE), design matrix, and contrast matrix in proper order. For this reason, <code>iSamples</code> is defined as <code><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames(TxSE)</a></code>, which should equal <code><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames(iDesign)</a></code>. Finally, <code><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames(iDesign)</a></code> should equal <code><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames(iContrasts)</a></code>.</p>
<p>As a fun matrix math review, you can print the crossproduct of the design and contrast matrices to view the samples used in each contrast.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/all">all</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(TxSE) <span class="op">==</span><span class="st"> </span>iSamples)
<span class="co">#&gt; [1] TRUE</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/all">all</a></span>(iSamples <span class="op">==</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(iDesign))
<span class="co">#&gt; [1] TRUE</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/all">all</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(iDesign) <span class="op">==</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(iContrasts))
<span class="co">#&gt; [1] TRUE</span>

iDesign <span class="op">%*%</span><span class="st"> </span>iContrasts;
<span class="co">#&gt;            Contrasts</span>
<span class="co">#&gt;             B-A</span>
<span class="co">#&gt;   ERR188297  -1</span>
<span class="co">#&gt;   ERR188088   1</span>
<span class="co">#&gt;   ERR188329  -1</span>
<span class="co">#&gt;   ERR188288   1</span>
<span class="co">#&gt;   ERR188021  -1</span>
<span class="co">#&gt;   ERR188356   1</span></code></pre></div>
</div>
<div id="rundiffsplice" class="section level4">
<h4 class="hasAnchor">
<a href="#rundiffsplice" class="anchor"></a>runDiffSplice()</h4>
<p>Given an expression data matrix, design matrix, and contrast matrix, <code><a href="../reference/runDiffSplice.html">splicejam::runDiffSplice()</a></code> can be called. Note that all abundance values of zero are set to <code>NA</code>, so they do not contribute to the analysis.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">iMatrixTxTPM &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class">assays</a></span>(TxSE[,iSamples])[[<span class="st">"abundance"</span>]];
iMatrixTxTPM[iMatrixTxTPM <span class="op">==</span><span class="st"> </span><span class="dv">0</span>] &lt;-<span class="st"> </span><span class="ot">NA</span>;
diffSpliceL &lt;-<span class="st"> </span><span class="kw"><a href="../reference/runDiffSplice.html">runDiffSplice</a></span>(
   <span class="dt">iMatrixTx=</span>iMatrixTxTPM,
   <span class="dt">detectedTx=</span>detectedTx,
   <span class="dt">tx2geneDF=</span>tx2geneDF,
   <span class="dt">txColname=</span>txColname,
   <span class="dt">geneColname=</span>geneColname,
   <span class="dt">iDesign=</span>iDesign,
   <span class="dt">iContrasts=</span>iContrasts,
   <span class="dt">cutoffFDR=</span><span class="fl">0.05</span>,
   <span class="dt">cutoffFold=</span><span class="fl">1.5</span>,
   <span class="dt">collapseByGene=</span><span class="ot">TRUE</span>,
   <span class="dt">spliceTest=</span><span class="st">"t"</span>,
   <span class="dt">verbose=</span><span class="ot">FALSE</span>,
   <span class="dt">useVoom=</span><span class="ot">FALSE</span>);
<span class="co">#&gt; Warning: Partial NA coefficients for 275 probe(s)</span>
<span class="co">#&gt; Total number of exons:  32645 </span>
<span class="co">#&gt; Total number of genes:  8702 </span>
<span class="co">#&gt; Number of genes with 1 exon:  0 </span>
<span class="co">#&gt; Mean number of exons in a gene:  4 </span>
<span class="co">#&gt; Max number of exons in a gene:  20</span>
diffSpliceHitsL &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">lapply</a></span>(diffSpliceL<span class="op">$</span>statsDFs, <span class="cf">function</span>(iDF){
   hitColname &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/jamba/topics/vigrep">vigrep</a></span>(<span class="st">"^hit "</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(iDF)), <span class="dv">1</span>);
   <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/character">as.character</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/subset">subset</a></span>(iDF, iDF[[hitColname]] <span class="op">!=</span><span class="st"> </span><span class="dv">0</span>)[,geneColname]);
});
<span class="kw"><a href="https://www.rdocumentation.org/packages/jamba/topics/printDebug">printDebug</a></span>(<span class="st">"Differential isoform hits per contrast:"</span>);
<span class="co">#&gt; ##  (17:14:23) 06Feb2019:  Differential isoform hits per contrast:</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/jamba/topics/sdim">sdim</a></span>(diffSpliceHitsL));
<span class="co">#&gt;     rows     class</span>
<span class="co">#&gt; B-A  661 character</span></code></pre></div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#r-setup">R setup</a></li>
      <li><a href="#salmon-quant-sf-gz-files">Salmon “quant.sf.gz” files</a></li>
      <li><a href="#transcript-to-gene-from-gtf">Transcript-to-gene from GTF</a></li>
      <li><a href="#import-salmon-transcript-data">Import Salmon transcript data</a></li>
      <li><a href="#import-salmon-gene-data-using-tx2genedf">Import Salmon gene data using tx2geneDF</a></li>
      <li><a href="#transcript-level-analysis-steps">Transcript-level analysis steps</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by James M. Ward.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
