<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Create a Sashimi Plot â€¢ splicejam</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Create a Sashimi Plot">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">splicejam</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.19.900</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/basic-rnaseq-workflow.html">Splicejam basic RNAseq workflow</a>
    </li>
    <li>
      <a href="../articles/create-a-sashimi-plot.html">Create a Sashimi Plot</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jmw86069/splicejam">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Create a Sashimi Plot</h1>
                        <h4 class="author">James M. Ward</h4>
            
            <h4 class="date">2019-03-26</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/jmw86069/splicejam/blob/master/vignettes/create-a-sashimi-plot.Rmd"><code>vignettes/create-a-sashimi-plot.Rmd</code></a></small>
      <div class="hidden name"><code>create-a-sashimi-plot.Rmd</code></div>

    </div>

    
    
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<p>This vignette is intended to describe how to create a Sashimi plot using RNA-seq data.</p>
<div id="requirements" class="section level3">
<h3 class="hasAnchor">
<a href="#requirements" class="anchor"></a>Requirements</h3>
<p>There are three basic requirements for a Sashimi plot:</p>
<ol style="list-style-type: decimal">
<li>Gene-exon structure</li>
<li>RNA-seq coverage data</li>
<li>exon-exon splice junction data</li>
</ol>
</div>
<div id="gene-exon-structure" class="section level3">
<h3 class="hasAnchor">
<a href="#gene-exon-structure" class="anchor"></a>Gene-exon structure</h3>
<p>There are several sources of gene-exon structures:</p>
<ul>
<li>
<p>Txdb transcript database, provided by <code><a href="https://www.rdocumentation.org/packages/GenomicFeatures/topics/TxDb-class">GenomicFeatures::TxDb</a></code></p>
<ul>
<li>It is convenient and recommended to use the same GTF or GFF file used in upstream RNA-seq processing steps as the source of exon data. In this case, you would use <code><a href="https://www.rdocumentation.org/packages/GenomicFeatures/topics/makeTxDbFromGFF">GenomicFeatures::makeTxDbFromGFF()</a></code> to create a <code>TxDb</code> object, then follow the steps starting with <code>TxDb</code>.</li>
<li>Bioconductor provides several Txdb data packages (see BiocView <a href="https://bioconductor.org/packages/release/BiocViews.html#___TxDb" class="uri">https://bioconductor.org/packages/release/BiocViews.html#___TxDb</a>) this they do not currently provide Gencode in TxDb format. For example, to install the package for hg19 UCSC knownGenes as a TxDb, use <code><a href="https://www.rdocumentation.org/packages/BiocManager/topics/install">BiocManager::install("TxDb.Hsapiens.UCSC.hg19.knownGene")</a></code>.</li>
</ul>
</li>
<li>
<p>GRanges or GRangesList objects, already assembled by other methods.</p>
<ul>
<li>For example <code><a href="https://www.rdocumentation.org/packages/rtracklayer/topics/export">rtracklayer::import()</a></code> has several methods that can import BED files, or GTF files.</li>
</ul>
</li>
</ul>
<p>The goal for Sashimi plots is to flatten exons to disjoint (non-overlapping) exons across the gene region of interest. The exons are used to determine the genomic regions to use for sequence coverage, and splice junction reads. The exons are also used to define intron region compression, in order to make the Sashimi plot more transcript-focused.</p>
<div id="starting-with-txdb" class="section level4">
<h4 class="hasAnchor">
<a href="#starting-with-txdb" class="anchor"></a>Starting with TxDb</h4>
<p>The recommended workflow involves assembling transcript exons, then aggregating detected transcripts into gene-exon models. The driving reason is that transcriptome data is increasingly comprehensive, and is derived from numerous tissue samples and cell types, many of which are not relevant nor representative of the transcriptome observed in a particular experimental condition. While it is not problematic to visualize a gene model whose exons are derived from all possible conditions, we found many gene models to be unnecessarily complex, therefore overshadowing the actual observed structure.</p>
<p>Said another way, many of the automated transcript structures are either just wrong, or were derived from conditions vastly different from those observed. Since tools like Salmon or Kallisto have the capability to determine detected and undetected transcripts based upon observing experimental data, we found it effective to use that information to define data-driven gene-exon models.</p>
<p>The example uses Bioconductor data package <code>TxDb.Mmusculus.UCSC.mm10.knownGene</code> using UCSC known genes for the mouse <code>mm10</code> genome assembly.</p>
<p>The <code>TxDb</code> objects do not permit assigning gene names nor gene symbols, so the example uses the Bioconductor annotation package <code>org.Mm.eg.db</code> to assign <code>gene_name</code> values. For any missing values, we use the format <code>LOC#</code> to indicate the Entrez gene ID, for example Entrez gene ID <code>1234</code> would be represented <code>LOC1234</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(splicejam);
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(jamba);
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(kableExtra)
<span class="cf">if</span> (<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">require</a></span>(TxDb.Mmusculus.UCSC.mm10.knownGene)) {

   <span class="co"># First obtain exons by transcript</span>
   exonsByTxMm10 &lt;-<span class="st"> </span><span class="kw">exonsBy</span>(TxDb.Mmusculus.UCSC.mm10.knownGene,
      <span class="dt">by=</span><span class="st">"tx"</span>,
      <span class="dt">use.names=</span><span class="ot">TRUE</span>);
   <span class="kw">values</span>(exonsByTxMm10<span class="op">@</span>unlistData)<span class="op">$</span>feature_type &lt;-<span class="st"> "exon"</span>;
   <span class="kw">values</span>(exonsByTxMm10<span class="op">@</span>unlistData)<span class="op">$</span>subclass &lt;-<span class="st"> "exon"</span>;
   
   <span class="co"># For added insight, obtain CDS exons by transcript (optional)</span>
   cdsByTxMm10 &lt;-<span class="st"> </span><span class="kw">cdsBy</span>(TxDb.Mmusculus.UCSC.mm10.knownGene,
      <span class="dt">by=</span><span class="st">"tx"</span>,
      <span class="dt">use.names=</span><span class="ot">TRUE</span>);
   <span class="kw">values</span>(cdsByTxMm10<span class="op">@</span>unlistData)<span class="op">$</span>feature_type &lt;-<span class="st"> "cds"</span>;
   
   <span class="co"># Now prepare tx_name, gene_id, gene_name data.frame,</span>
   <span class="co"># surprisingly difficult</span>
   tx2geneMm10 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/message">suppressMessages</a></span>(
      AnnotationDbi<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/AnnotationDbi/topics/AnnotationDb-class">select</a></span>(TxDb.Mmusculus.UCSC.mm10.knownGene,
         <span class="kw">keys</span>(TxDb.Mmusculus.UCSC.mm10.knownGene, <span class="st">"GENEID"</span>),
         <span class="dt">columns=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"GENEID"</span>,<span class="st">"TXNAME"</span>),
         <span class="dt">keytype=</span><span class="st">"GENEID"</span>
      )
   );
   tx2geneMm10 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/jamba/topics/renameColumn">renameColumn</a></span>(tx2geneMm10,
      <span class="dt">from=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"GENEID"</span>, <span class="st">"TXNAME"</span>, <span class="st">"TXTYPE"</span>),
      <span class="dt">to=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"gene_id"</span>, <span class="st">"transcript_id"</span>, <span class="st">"transcript_type"</span>));

   <span class="co"># add gene_name using org.Mm.eg.db</span>
   <span class="cf">if</span> (<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">require</a></span>(org.Mm.eg.db)) {
      gene_ids &lt;-<span class="st"> </span><span class="kw">values</span>(<span class="kw">genes</span>(TxDb.Mmusculus.UCSC.mm10.knownGene))<span class="op">$</span>gene_id;
      gene_namesL &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/get">mget</a></span>(gene_ids,
         org.Mm.egSYMBOL,
         <span class="dt">ifnotfound=</span><span class="ot">NA</span>);
      ## Convert list to vector taking the first gene_name each
      gene_names &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unlist">unlist</a></span>(<span class="kw">heads</span>(S4Vectors<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/S4Vectors/topics/List-class">List</a></span>(gene_namesL), <span class="dv">1</span>));
      ## Replace NA with LOC# format
      <span class="cf">if</span> (<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/any">any</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/NA">is.na</a></span>(gene_names))) {
         gene_na &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/which">which</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/NA">is.na</a></span>(gene_names));
         gene_names[gene_na] &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(<span class="st">"LOC"</span>,
            <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(gene_names[gene_na]));
      }
      tx2geneMm10<span class="op">$</span>gene_name &lt;-<span class="st"> </span>gene_names[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/character">as.character</a></span>(tx2geneMm10<span class="op">$</span>gene_id)];
   } <span class="cf">else</span> {
      tx2geneMm10<span class="op">$</span>gene_name &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/character">as.character</a></span>(tx2geneMm10<span class="op">$</span>gene_id);
   }
   <span class="co"># print the first 20 rows to show the content</span>
   <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(tx2geneMm10, <span class="dv">20</span>));
   
   <span class="co"># flatten exons to the gene level</span>
   <span class="co"># for speed, we will only process "Gria1", and "Ntrk2"</span>
   flatExonsByGeneMm10 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/flattenExonsByGene.html">flattenExonsByGene</a></span>(<span class="dt">exonsByTx=</span>exonsByTxMm10,
      <span class="dt">cdsByTx=</span>cdsByTxMm10,
      <span class="dt">genes=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"Gria1"</span>, <span class="st">"Ntrk2"</span>),
      <span class="dt">tx2geneDF=</span>tx2geneMm10,
      <span class="dt">verbose=</span><span class="ot">FALSE</span>);

   <span class="co"># to be fancy, also flatten transcripts, to include CDS ranges   </span>
   flatExonsByTxMm10 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/flattenExonsByGene.html">flattenExonsByGene</a></span>(<span class="dt">exonsByTx=</span>exonsByTxMm10,
      <span class="dt">cdsByTx=</span>cdsByTxMm10,
      <span class="dt">tx2geneDF=</span>tx2geneMm10,
      <span class="dt">by=</span><span class="st">"tx"</span>,
      <span class="dt">genes=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"Gria1"</span>, <span class="st">"Ntrk2"</span>));
}
<span class="co">#&gt; Loading required package: TxDb.Mmusculus.UCSC.mm10.knownGene</span>
<span class="co">#&gt; Loading required package: GenomicFeatures</span>
<span class="co">#&gt; Warning: package 'GenomicFeatures' was built under R version 3.5.2</span>
<span class="co">#&gt; Loading required package: BiocGenerics</span>
<span class="co">#&gt; Loading required package: parallel</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'BiocGenerics'</span>
<span class="co">#&gt; The following objects are masked from 'package:parallel':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,</span>
<span class="co">#&gt;     clusterExport, clusterMap, parApply, parCapply, parLapply,</span>
<span class="co">#&gt;     parLapplyLB, parRapply, parSapply, parSapplyLB</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     IQR, mad, sd, var, xtabs</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     Filter, Find, Map, Position, Reduce, anyDuplicated, append,</span>
<span class="co">#&gt;     as.data.frame, basename, cbind, colMeans, colSums, colnames,</span>
<span class="co">#&gt;     dirname, do.call, duplicated, eval, evalq, get, grep, grepl,</span>
<span class="co">#&gt;     intersect, is.unsorted, lapply, lengths, mapply, match, mget,</span>
<span class="co">#&gt;     order, paste, pmax, pmax.int, pmin, pmin.int, rank, rbind,</span>
<span class="co">#&gt;     rowMeans, rowSums, rownames, sapply, setdiff, sort, table,</span>
<span class="co">#&gt;     tapply, union, unique, unsplit, which, which.max, which.min</span>
<span class="co">#&gt; Loading required package: S4Vectors</span>
<span class="co">#&gt; Loading required package: stats4</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'S4Vectors'</span>
<span class="co">#&gt; The following object is masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     expand.grid</span>
<span class="co">#&gt; Loading required package: IRanges</span>
<span class="co">#&gt; Loading required package: GenomeInfoDb</span>
<span class="co">#&gt; Loading required package: GenomicRanges</span>
<span class="co">#&gt; Loading required package: AnnotationDbi</span>
<span class="co">#&gt; Loading required package: Biobase</span>
<span class="co">#&gt; Welcome to Bioconductor</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     Vignettes contain introductory material; view with</span>
<span class="co">#&gt;     'browseVignettes()'. To cite Bioconductor, see</span>
<span class="co">#&gt;     'citation("Biobase")', and for packages 'citation("pkgname")'.</span>
<span class="co">#&gt; Loading required package: org.Mm.eg.db</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;      gene_id transcript_id     gene_name</span>
<span class="co">#&gt; 1  100009600    uc009veu.1         Zglp1</span>
<span class="co">#&gt; 2  100009600    uc033jjg.1         Zglp1</span>
<span class="co">#&gt; 3  100009609    uc012fog.1       Vmn2r65</span>
<span class="co">#&gt; 4  100009614    uc011xhj.2       Gm10024</span>
<span class="co">#&gt; 5  100009664    uc007inp.3 F630206G17Rik</span>
<span class="co">#&gt; 6     100012    uc008vqx.2          Oog3</span>
<span class="co">#&gt; 7     100017    uc008vfo.3       Ldlrap1</span>
<span class="co">#&gt; 8     100017    uc012dmx.1       Ldlrap1</span>
<span class="co">#&gt; 9     100019    uc008sey.1          Mdn1</span>
<span class="co">#&gt; 10    100019    uc008sez.1          Mdn1</span>
<span class="co">#&gt; 11    100019    uc008sfa.1          Mdn1</span>
<span class="co">#&gt; 12    100019    uc008sfb.1          Mdn1</span>
<span class="co">#&gt; 13    100019    uc008sfc.1          Mdn1</span>
<span class="co">#&gt; 14    100019    uc008sfd.1          Mdn1</span>
<span class="co">#&gt; 15 100033459    uc007drs.1        Ifi208</span>
<span class="co">#&gt; 16 100033459    uc007drt.1        Ifi208</span>
<span class="co">#&gt; 17 100034251    uc007kpq.1        Wfdc17</span>
<span class="co">#&gt; 18 100034361    uc008lzg.1        Mfap1b</span>
<span class="co">#&gt; 19 100034363    uc009ujg.2      Tmsb15b2</span>
<span class="co">#&gt; 20 100034684    uc007zcm.1      BC100530</span></code></pre></div>
<p>Once the gene-transcript-exon data is prepared, the process of plotting data for one gene is demonstrated below.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="cf">if</span> (<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">require</a></span>(TxDb.Mmusculus.UCSC.mm10.knownGene)) {
   <span class="co"># Pull out Gria1</span>
   grlGria1 &lt;-<span class="st"> </span>flatExonsByGeneMm10[[<span class="st">"Gria1"</span>]];

   ## Plot a basic gene-exon structure
   ggGria1exons &lt;-<span class="st"> </span><span class="kw"><a href="../reference/gene2gg.html">gene2gg</a></span>(<span class="dt">gene=</span><span class="st">"Gria1"</span>,
      <span class="dt">flatExonsByGene=</span>flatExonsByGeneMm10,
      <span class="dt">exonLabelSize=</span><span class="dv">6</span>);
   <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(ggGria1exons);

   ## Compare to the gene structure without compressing introns
   gg1full &lt;-<span class="st"> </span><span class="kw"><a href="../reference/gene2gg.html">gene2gg</a></span>(<span class="dt">gene=</span><span class="st">"Gria1"</span>,
      <span class="dt">flatExonsByGene=</span>flatExonsByGeneMm10,
      <span class="dt">compressGaps=</span><span class="ot">FALSE</span>)
   <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(gg1full);
   
   ## Plot a slightly more detailed gene-transcript-exon structure
   ggGria1exonsTx &lt;-<span class="st"> </span><span class="kw"><a href="../reference/gene2gg.html">gene2gg</a></span>(<span class="dt">gene=</span><span class="st">"Gria1"</span>,
      <span class="dt">flatExonsByGene=</span>flatExonsByGeneMm10,
      <span class="dt">flatExonsByTx=</span>flatExonsByTxMm10,
      <span class="dt">tx2geneDF=</span>tx2geneMm10);
   <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(ggGria1exonsTx);
   
   ## Notice how difficult it is to see exon15 and exon16 are
   ## mutually exclusive exons
   gg2full &lt;-<span class="st"> </span><span class="kw"><a href="../reference/gene2gg.html">gene2gg</a></span>(<span class="dt">gene=</span><span class="st">"Gria1"</span>,
      <span class="dt">flatExonsByGene=</span>flatExonsByGeneMm10,
      <span class="dt">flatExonsByTx=</span>flatExonsByTxMm10,
      <span class="dt">tx2geneDF=</span>tx2geneMm10,
      <span class="dt">compressGaps=</span><span class="ot">FALSE</span>);
   <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(gg2full);

}
<span class="co">#&gt; Warning: package 'colorspace' was built under R version 3.5.2</span>
<span class="co">#&gt; ##  (15:36:54) 26Mar2019:  head(df): </span>
<span class="co">#&gt;          x     y id grl_name      gr_name gene_name subclass gene_nameExon</span>
<span class="co">#&gt; 1 57011618 -0.15  1    Gria1 Gria1_exon1a     Gria1   noncds  Gria1_exon1a</span>
<span class="co">#&gt; 2 57012033 -0.15  1    Gria1 Gria1_exon1a     Gria1   noncds  Gria1_exon1a</span>
<span class="co">#&gt; 3 57012033  0.15  1    Gria1 Gria1_exon1a     Gria1   noncds  Gria1_exon1a</span>
<span class="co">#&gt; 4 57011618  0.15  1    Gria1 Gria1_exon1a     Gria1   noncds  Gria1_exon1a</span>
<span class="co">#&gt; 5 57012034 -0.30  2    Gria1 Gria1_exon1b     Gria1      cds  Gria1_exon1b</span>
<span class="co">#&gt; 6 57012115 -0.30  2    Gria1 Gria1_exon1b     Gria1      cds  Gria1_exon1b</span>
<span class="co">#&gt;   feature_type</span>
<span class="co">#&gt; 1         exon</span>
<span class="co">#&gt; 2         exon</span>
<span class="co">#&gt; 3         exon</span>
<span class="co">#&gt; 4         exon</span>
<span class="co">#&gt; 5         exon</span>
<span class="co">#&gt; 6         exon</span>
<span class="co">#&gt;            x     y id grl_name       gr_name gene_name subclass</span>
<span class="co">#&gt; 143 57327646  0.30 36    Gria1 Gria1_exon18a     Gria1      cds</span>
<span class="co">#&gt; 144 57327443  0.30 36    Gria1 Gria1_exon18a     Gria1      cds</span>
<span class="co">#&gt; 145 57327647 -0.15 37    Gria1 Gria1_exon18b     Gria1   noncds</span>
<span class="co">#&gt; 146 57330244 -0.15 37    Gria1 Gria1_exon18b     Gria1   noncds</span>
<span class="co">#&gt; 147 57330244  0.15 37    Gria1 Gria1_exon18b     Gria1   noncds</span>
<span class="co">#&gt; 148 57327647  0.15 37    Gria1 Gria1_exon18b     Gria1   noncds</span>
<span class="co">#&gt;     gene_nameExon feature_type</span>
<span class="co">#&gt; 143 Gria1_exon18a         exon</span>
<span class="co">#&gt; 144 Gria1_exon18a         exon</span>
<span class="co">#&gt; 145 Gria1_exon18b         exon</span>
<span class="co">#&gt; 146 Gria1_exon18b         exon</span>
<span class="co">#&gt; 147 Gria1_exon18b         exon</span>
<span class="co">#&gt; 148 Gria1_exon18b         exon</span>
<span class="co">#&gt; Warning: package 'data.table' was built under R version 3.5.2</span>
<span class="co">#&gt; Warning: Ignoring unknown parameters: fill</span>
<span class="co">#&gt; ##  (15:36:55) 26Mar2019:  head(df): </span>
<span class="co">#&gt;          x     y id grl_name      gr_name gene_name subclass gene_nameExon</span>
<span class="co">#&gt; 1 57011618 -0.15  1    Gria1 Gria1_exon1a     Gria1   noncds  Gria1_exon1a</span>
<span class="co">#&gt; 2 57012033 -0.15  1    Gria1 Gria1_exon1a     Gria1   noncds  Gria1_exon1a</span>
<span class="co">#&gt; 3 57012033  0.15  1    Gria1 Gria1_exon1a     Gria1   noncds  Gria1_exon1a</span>
<span class="co">#&gt; 4 57011618  0.15  1    Gria1 Gria1_exon1a     Gria1   noncds  Gria1_exon1a</span>
<span class="co">#&gt; 5 57012034 -0.30  2    Gria1 Gria1_exon1b     Gria1      cds  Gria1_exon1b</span>
<span class="co">#&gt; 6 57012115 -0.30  2    Gria1 Gria1_exon1b     Gria1      cds  Gria1_exon1b</span>
<span class="co">#&gt;   feature_type</span>
<span class="co">#&gt; 1         exon</span>
<span class="co">#&gt; 2         exon</span>
<span class="co">#&gt; 3         exon</span>
<span class="co">#&gt; 4         exon</span>
<span class="co">#&gt; 5         exon</span>
<span class="co">#&gt; 6         exon</span>
<span class="co">#&gt;            x     y id grl_name       gr_name gene_name subclass</span>
<span class="co">#&gt; 143 57327646  0.30 36    Gria1 Gria1_exon18a     Gria1      cds</span>
<span class="co">#&gt; 144 57327443  0.30 36    Gria1 Gria1_exon18a     Gria1      cds</span>
<span class="co">#&gt; 145 57327647 -0.15 37    Gria1 Gria1_exon18b     Gria1   noncds</span>
<span class="co">#&gt; 146 57330244 -0.15 37    Gria1 Gria1_exon18b     Gria1   noncds</span>
<span class="co">#&gt; 147 57330244  0.15 37    Gria1 Gria1_exon18b     Gria1   noncds</span>
<span class="co">#&gt; 148 57327647  0.15 37    Gria1 Gria1_exon18b     Gria1   noncds</span>
<span class="co">#&gt;     gene_nameExon feature_type</span>
<span class="co">#&gt; 143 Gria1_exon18a         exon</span>
<span class="co">#&gt; 144 Gria1_exon18a         exon</span>
<span class="co">#&gt; 145 Gria1_exon18b         exon</span>
<span class="co">#&gt; 146 Gria1_exon18b         exon</span>
<span class="co">#&gt; 147 Gria1_exon18b         exon</span>
<span class="co">#&gt; 148 Gria1_exon18b         exon</span>
<span class="co">#&gt; Warning: Ignoring unknown parameters: fill</span></code></pre></div>
<p><img src="create-a-sashimi-plot_files/figure-html/gene_plot-1.png" width="672"></p>
<pre><code>#&gt; ##  (15:36:56) 26Mar2019:  Checking flatExonsByTx. 
#&gt; ##  (15:36:56) 26Mar2019:  head(df): 
#&gt;          x     y id   grl_name           gr_name transcript_id gene_name
#&gt; 1 57011618 -0.15  1 uc007izr.3 uc007izr.3_exon1a    uc007izr.3     Gria1
#&gt; 2 57012033 -0.15  1 uc007izr.3 uc007izr.3_exon1a    uc007izr.3     Gria1
#&gt; 3 57012033  0.15  1 uc007izr.3 uc007izr.3_exon1a    uc007izr.3     Gria1
#&gt; 4 57011618  0.15  1 uc007izr.3 uc007izr.3_exon1a    uc007izr.3     Gria1
#&gt; 5 57012034 -0.30  2 uc007izr.3 uc007izr.3_exon1b    uc007izr.3     Gria1
#&gt; 6 57012115 -0.30  2 uc007izr.3 uc007izr.3_exon1b    uc007izr.3     Gria1
#&gt;   subclass transcript_idExon feature_type gene_nameExon
#&gt; 1   noncds uc007izr.3_exon1a         exon          &lt;NA&gt;
#&gt; 2   noncds uc007izr.3_exon1a         exon          &lt;NA&gt;
#&gt; 3   noncds uc007izr.3_exon1a         exon          &lt;NA&gt;
#&gt; 4   noncds uc007izr.3_exon1a         exon          &lt;NA&gt;
#&gt; 5      cds uc007izr.3_exon1b         exon          &lt;NA&gt;
#&gt; 6      cds uc007izr.3_exon1b         exon          &lt;NA&gt;
#&gt;            x    y  id grl_name       gr_name transcript_id gene_name
#&gt; 575 57327646 4.30 144    Gria1 Gria1_exon18a          &lt;NA&gt;     Gria1
#&gt; 576 57327443 4.30 144    Gria1 Gria1_exon18a          &lt;NA&gt;     Gria1
#&gt; 577 57327647 3.85 145    Gria1 Gria1_exon18b          &lt;NA&gt;     Gria1
#&gt; 578 57330244 3.85 145    Gria1 Gria1_exon18b          &lt;NA&gt;     Gria1
#&gt; 579 57330244 4.15 145    Gria1 Gria1_exon18b          &lt;NA&gt;     Gria1
#&gt; 580 57327647 4.15 145    Gria1 Gria1_exon18b          &lt;NA&gt;     Gria1
#&gt;     subclass transcript_idExon feature_type gene_nameExon
#&gt; 575      cds              &lt;NA&gt;         exon Gria1_exon18a
#&gt; 576      cds              &lt;NA&gt;         exon Gria1_exon18a
#&gt; 577   noncds              &lt;NA&gt;         exon Gria1_exon18b
#&gt; 578   noncds              &lt;NA&gt;         exon Gria1_exon18b
#&gt; 579   noncds              &lt;NA&gt;         exon Gria1_exon18b
#&gt; 580   noncds              &lt;NA&gt;         exon Gria1_exon18b
#&gt; Warning: Ignoring unknown parameters: fill</code></pre>
<p><img src="create-a-sashimi-plot_files/figure-html/gene_plot-2.png" width="672"></p>
<pre><code>#&gt; Warning: Removed 1 rows containing missing values (geom_text_repel).
#&gt; ##  (15:36:57) 26Mar2019:  Checking flatExonsByTx. 
#&gt; ##  (15:36:57) 26Mar2019:  head(df): 
#&gt;          x     y id   grl_name           gr_name transcript_id gene_name
#&gt; 1 57011618 -0.15  1 uc007izr.3 uc007izr.3_exon1a    uc007izr.3     Gria1
#&gt; 2 57012033 -0.15  1 uc007izr.3 uc007izr.3_exon1a    uc007izr.3     Gria1
#&gt; 3 57012033  0.15  1 uc007izr.3 uc007izr.3_exon1a    uc007izr.3     Gria1
#&gt; 4 57011618  0.15  1 uc007izr.3 uc007izr.3_exon1a    uc007izr.3     Gria1
#&gt; 5 57012034 -0.30  2 uc007izr.3 uc007izr.3_exon1b    uc007izr.3     Gria1
#&gt; 6 57012115 -0.30  2 uc007izr.3 uc007izr.3_exon1b    uc007izr.3     Gria1
#&gt;   subclass transcript_idExon feature_type gene_nameExon
#&gt; 1   noncds uc007izr.3_exon1a         exon          &lt;NA&gt;
#&gt; 2   noncds uc007izr.3_exon1a         exon          &lt;NA&gt;
#&gt; 3   noncds uc007izr.3_exon1a         exon          &lt;NA&gt;
#&gt; 4   noncds uc007izr.3_exon1a         exon          &lt;NA&gt;
#&gt; 5      cds uc007izr.3_exon1b         exon          &lt;NA&gt;
#&gt; 6      cds uc007izr.3_exon1b         exon          &lt;NA&gt;
#&gt;            x    y  id grl_name       gr_name transcript_id gene_name
#&gt; 575 57327646 4.30 144    Gria1 Gria1_exon18a          &lt;NA&gt;     Gria1
#&gt; 576 57327443 4.30 144    Gria1 Gria1_exon18a          &lt;NA&gt;     Gria1
#&gt; 577 57327647 3.85 145    Gria1 Gria1_exon18b          &lt;NA&gt;     Gria1
#&gt; 578 57330244 3.85 145    Gria1 Gria1_exon18b          &lt;NA&gt;     Gria1
#&gt; 579 57330244 4.15 145    Gria1 Gria1_exon18b          &lt;NA&gt;     Gria1
#&gt; 580 57327647 4.15 145    Gria1 Gria1_exon18b          &lt;NA&gt;     Gria1
#&gt;     subclass transcript_idExon feature_type gene_nameExon
#&gt; 575      cds              &lt;NA&gt;         exon Gria1_exon18a
#&gt; 576      cds              &lt;NA&gt;         exon Gria1_exon18a
#&gt; 577   noncds              &lt;NA&gt;         exon Gria1_exon18b
#&gt; 578   noncds              &lt;NA&gt;         exon Gria1_exon18b
#&gt; 579   noncds              &lt;NA&gt;         exon Gria1_exon18b
#&gt; 580   noncds              &lt;NA&gt;         exon Gria1_exon18b
#&gt; Warning: Ignoring unknown parameters: fill</code></pre>
<p><img src="create-a-sashimi-plot_files/figure-html/gene_plot-3.png" width="672"></p>
<pre><code>#&gt; Warning: Removed 1 rows containing missing values (geom_text_repel).</code></pre>
<p><img src="create-a-sashimi-plot_files/figure-html/gene_plot-4.png" width="672"></p>
</div>
</div>
<div id="coverage-and-junction-data" class="section level3">
<h3 class="hasAnchor">
<a href="#coverage-and-junction-data" class="anchor"></a>Coverage and Junction data</h3>
<p>The current workflow is intended to use bigWig files as the primary source for RNA-seq coverage data.</p>
<p>To provide bigWig coverage files to the <code><a href="../reference/prepareSashimi.html">prepareSashimi()</a></code> function, requires a <code>data.frame</code> with column names:</p>
<ul>
<li>
<code>"url"</code> - the path to a local file, or a valid web URL.</li>
<li>
<code>"type"</code> - the type of file, currently allowing <code>"bw"</code> for bigWig data, or <code>"junction"</code> for a BED12 formatted file containing splice junction reads.</li>
<li>
<code>"sample_id"</code> - a unique biological sample identifier, which can be repeated for replicate files as needed.</li>
<li>
<code>"scale_factor"</code> - optional column with numeric values which are multiplied by the coverage (or junction read counts), in order to define normalized values for plotting purposes.</li>
</ul>
<p>A bigWig coverage file can only store coverage for one strand, when using strand-specific RNA-seq. To provide negative strand data, either provide bigWig coverage scores which are negative values (never above zero, and containing at least one negative value), or supply a negative value for the <code>"scale_factor"</code> field as described above.</p>
<p>The example below defines a <code>data.frame</code> named <code>filesDF</code> which contains bigWig coverage, and splice junction data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## assemble a data.frame
baseurl &lt;-<span class="st"> "https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/"</span>;

<span class="co"># BED files with junction reads</span>
bedext &lt;-<span class="st"> ".STAR_mm10.combinedJunctions.bed"</span>;
bwext &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"492_1.sickle.merged.cutadapt.STAR_mm10.pos.bw"</span>,
   <span class="st">"492_1.sickle.merged.cutadapt.STAR_mm10.neg.bw"</span>);
c1 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"CA1"</span>, <span class="st">"CA2"</span>);
r1 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"CB"</span>, <span class="st">"DE"</span>);
bedsamples &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(c1, <span class="dt">each=</span><span class="dv">2</span>), <span class="st">"_"</span>, r1);
bedurls &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(baseurl,
   bedsamples,
   bedext);

<span class="co"># bigWig files with strand-specific read coverage</span>
bwsamples &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(c1, <span class="dt">each=</span><span class="dv">4</span>),
   <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(r1, <span class="dt">each=</span><span class="dv">2</span>));
bwsamples1 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(c1, <span class="dt">each=</span><span class="dv">4</span>),
   <span class="st">"_"</span>,
   <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(r1, <span class="dt">each=</span><span class="dv">2</span>));
bwurls &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(baseurl, <span class="st">"NS50211"</span>,
   bwsamples,
   bwext);

<span class="co"># Assemble into a data.frame</span>
filesDF &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>(<span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>,
   <span class="dt">check.names=</span><span class="ot">FALSE</span>,
   <span class="dt">url=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(bedurls, bwurls),
   <span class="dt">type=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"junction"</span>, <span class="st">"bw"</span>), <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">4</span>,<span class="dv">8</span>)),
   <span class="dt">sample_id=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(bedsamples, bwsamples1),
   <span class="dt">scale_factor=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>,<span class="dv">3</span>), <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(bedurls), <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(bwurls))));
filesDF;
<span class="co">#&gt;                                                                                                             url</span>
<span class="co">#&gt; 1                     https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/CA1_CB.STAR_mm10.combinedJunctions.bed</span>
<span class="co">#&gt; 2                     https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/CA1_DE.STAR_mm10.combinedJunctions.bed</span>
<span class="co">#&gt; 3                     https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/CA2_CB.STAR_mm10.combinedJunctions.bed</span>
<span class="co">#&gt; 4                     https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/CA2_DE.STAR_mm10.combinedJunctions.bed</span>
<span class="co">#&gt; 5  https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/NS50211CA1CB492_1.sickle.merged.cutadapt.STAR_mm10.pos.bw</span>
<span class="co">#&gt; 6  https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/NS50211CA1CB492_1.sickle.merged.cutadapt.STAR_mm10.neg.bw</span>
<span class="co">#&gt; 7  https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/NS50211CA1DE492_1.sickle.merged.cutadapt.STAR_mm10.pos.bw</span>
<span class="co">#&gt; 8  https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/NS50211CA1DE492_1.sickle.merged.cutadapt.STAR_mm10.neg.bw</span>
<span class="co">#&gt; 9  https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/NS50211CA2CB492_1.sickle.merged.cutadapt.STAR_mm10.pos.bw</span>
<span class="co">#&gt; 10 https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/NS50211CA2CB492_1.sickle.merged.cutadapt.STAR_mm10.neg.bw</span>
<span class="co">#&gt; 11 https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/NS50211CA2DE492_1.sickle.merged.cutadapt.STAR_mm10.pos.bw</span>
<span class="co">#&gt; 12 https://orio.niehs.nih.gov/ucscview/farrisHub/mm10/NS50211CA2DE492_1.sickle.merged.cutadapt.STAR_mm10.neg.bw</span>
<span class="co">#&gt;        type sample_id scale_factor</span>
<span class="co">#&gt; 1  junction    CA1_CB            1</span>
<span class="co">#&gt; 2  junction    CA1_DE            1</span>
<span class="co">#&gt; 3  junction    CA2_CB            1</span>
<span class="co">#&gt; 4  junction    CA2_DE            1</span>
<span class="co">#&gt; 5        bw    CA1_CB            3</span>
<span class="co">#&gt; 6        bw    CA1_CB            3</span>
<span class="co">#&gt; 7        bw    CA1_DE            3</span>
<span class="co">#&gt; 8        bw    CA1_DE            3</span>
<span class="co">#&gt; 9        bw    CA2_CB            3</span>
<span class="co">#&gt; 10       bw    CA2_CB            3</span>
<span class="co">#&gt; 11       bw    CA2_DE            3</span>
<span class="co">#&gt; 12       bw    CA2_DE            3</span></code></pre></div>
</div>
<div id="prepare-sashimi-data" class="section level3">
<h3 class="hasAnchor">
<a href="#prepare-sashimi-data" class="anchor"></a>Prepare Sashimi Data</h3>
<p>The main function <code><a href="../reference/prepareSashimi.html">prepareSashimi()</a></code> is intended to assemble the sources of data together into one list of <code>data.frame</code> results which can be used to produce ggplot2 visualizations.</p>
<p>For this example, we use the Gria1 gene in mouse mm10.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">shGria1 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/prepareSashimi.html">prepareSashimi</a></span>(<span class="dt">gene=</span><span class="st">"Gria1"</span>,
   <span class="dt">flatExonsByGene=</span>flatExonsByGeneMm10,
   <span class="dt">minJunctionScore=</span><span class="dv">100</span>,
   <span class="dt">sample_id=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"CA1_CB"</span>, <span class="st">"CA2_CB"</span>),
   <span class="dt">filesDF=</span>filesDF)
<span class="co">#&gt; Warning in .local(con, format, text, ...): 'which' contains seqlevels</span>
<span class="co">#&gt; not known to BigWig file: chr1_GL456210_random, chr1_GL456211_random,</span>
<span class="co">#&gt; chr1_GL456212_random, chr1_GL456213_random, chr1_GL456221_random,</span>
<span class="co">#&gt; chr4_GL456216_random, chr4_GL456350_random, chr4_JH584292_random,</span>
<span class="co">#&gt; chr4_JH584293_random, chr4_JH584294_random, chr4_JH584295_random,</span>
<span class="co">#&gt; chr5_GL456354_random, chr5_JH584296_random, chr5_JH584297_random,</span>
<span class="co">#&gt; chr5_JH584298_random, chr5_JH584299_random, chr7_GL456219_random,</span>
<span class="co">#&gt; chrX_GL456233_random, chrY_JH584300_random, chrY_JH584301_random,</span>
<span class="co">#&gt; chrY_JH584302_random, chrY_JH584303_random, chrUn_GL456239, chrUn_GL456359,</span>
<span class="co">#&gt; chrUn_GL456360, chrUn_GL456366, chrUn_GL456367, chrUn_GL456368,</span>
<span class="co">#&gt; chrUn_GL456370, chrUn_GL456372, chrUn_GL456378, chrUn_GL456379,</span>
<span class="co">#&gt; chrUn_GL456381, chrUn_GL456382, chrUn_GL456383, chrUn_GL456385,</span>
<span class="co">#&gt; chrUn_GL456387, chrUn_GL456389, chrUn_GL456390, chrUn_GL456392,</span>
<span class="co">#&gt; chrUn_GL456393, chrUn_GL456394, chrUn_GL456396, chrUn_JH584304</span>

<span class="co">#&gt; Warning in .local(con, format, text, ...): 'which' contains seqlevels</span>
<span class="co">#&gt; not known to BigWig file: chr1_GL456210_random, chr1_GL456211_random,</span>
<span class="co">#&gt; chr1_GL456212_random, chr1_GL456213_random, chr1_GL456221_random,</span>
<span class="co">#&gt; chr4_GL456216_random, chr4_GL456350_random, chr4_JH584292_random,</span>
<span class="co">#&gt; chr4_JH584293_random, chr4_JH584294_random, chr4_JH584295_random,</span>
<span class="co">#&gt; chr5_GL456354_random, chr5_JH584296_random, chr5_JH584297_random,</span>
<span class="co">#&gt; chr5_JH584298_random, chr5_JH584299_random, chr7_GL456219_random,</span>
<span class="co">#&gt; chrX_GL456233_random, chrY_JH584300_random, chrY_JH584301_random,</span>
<span class="co">#&gt; chrY_JH584302_random, chrY_JH584303_random, chrUn_GL456239, chrUn_GL456359,</span>
<span class="co">#&gt; chrUn_GL456360, chrUn_GL456366, chrUn_GL456367, chrUn_GL456368,</span>
<span class="co">#&gt; chrUn_GL456370, chrUn_GL456372, chrUn_GL456378, chrUn_GL456379,</span>
<span class="co">#&gt; chrUn_GL456381, chrUn_GL456382, chrUn_GL456383, chrUn_GL456385,</span>
<span class="co">#&gt; chrUn_GL456387, chrUn_GL456389, chrUn_GL456390, chrUn_GL456392,</span>
<span class="co">#&gt; chrUn_GL456393, chrUn_GL456394, chrUn_GL456396, chrUn_JH584304</span>

<span class="co">#&gt; Warning in .local(con, format, text, ...): 'which' contains seqlevels</span>
<span class="co">#&gt; not known to BigWig file: chr1_GL456210_random, chr1_GL456211_random,</span>
<span class="co">#&gt; chr1_GL456212_random, chr1_GL456213_random, chr1_GL456221_random,</span>
<span class="co">#&gt; chr4_GL456216_random, chr4_GL456350_random, chr4_JH584292_random,</span>
<span class="co">#&gt; chr4_JH584293_random, chr4_JH584294_random, chr4_JH584295_random,</span>
<span class="co">#&gt; chr5_GL456354_random, chr5_JH584296_random, chr5_JH584297_random,</span>
<span class="co">#&gt; chr5_JH584298_random, chr5_JH584299_random, chr7_GL456219_random,</span>
<span class="co">#&gt; chrX_GL456233_random, chrY_JH584300_random, chrY_JH584301_random,</span>
<span class="co">#&gt; chrY_JH584302_random, chrY_JH584303_random, chrUn_GL456239, chrUn_GL456359,</span>
<span class="co">#&gt; chrUn_GL456360, chrUn_GL456366, chrUn_GL456367, chrUn_GL456368,</span>
<span class="co">#&gt; chrUn_GL456370, chrUn_GL456372, chrUn_GL456378, chrUn_GL456379,</span>
<span class="co">#&gt; chrUn_GL456381, chrUn_GL456382, chrUn_GL456383, chrUn_GL456385,</span>
<span class="co">#&gt; chrUn_GL456387, chrUn_GL456389, chrUn_GL456390, chrUn_GL456392,</span>
<span class="co">#&gt; chrUn_GL456393, chrUn_GL456394, chrUn_GL456396, chrUn_JH584304</span>

<span class="co">#&gt; Warning in .local(con, format, text, ...): 'which' contains seqlevels</span>
<span class="co">#&gt; not known to BigWig file: chr1_GL456210_random, chr1_GL456211_random,</span>
<span class="co">#&gt; chr1_GL456212_random, chr1_GL456213_random, chr1_GL456221_random,</span>
<span class="co">#&gt; chr4_GL456216_random, chr4_GL456350_random, chr4_JH584292_random,</span>
<span class="co">#&gt; chr4_JH584293_random, chr4_JH584294_random, chr4_JH584295_random,</span>
<span class="co">#&gt; chr5_GL456354_random, chr5_JH584296_random, chr5_JH584297_random,</span>
<span class="co">#&gt; chr5_JH584298_random, chr5_JH584299_random, chr7_GL456219_random,</span>
<span class="co">#&gt; chrX_GL456233_random, chrY_JH584300_random, chrY_JH584301_random,</span>
<span class="co">#&gt; chrY_JH584302_random, chrY_JH584303_random, chrUn_GL456239, chrUn_GL456359,</span>
<span class="co">#&gt; chrUn_GL456360, chrUn_GL456366, chrUn_GL456367, chrUn_GL456368,</span>
<span class="co">#&gt; chrUn_GL456370, chrUn_GL456372, chrUn_GL456378, chrUn_GL456379,</span>
<span class="co">#&gt; chrUn_GL456381, chrUn_GL456382, chrUn_GL456383, chrUn_GL456385,</span>
<span class="co">#&gt; chrUn_GL456387, chrUn_GL456389, chrUn_GL456390, chrUn_GL456392,</span>
<span class="co">#&gt; chrUn_GL456393, chrUn_GL456394, chrUn_GL456396, chrUn_JH584304</span></code></pre></div>
</div>
<div id="plot-sashimi-data" class="section level3">
<h3 class="hasAnchor">
<a href="#plot-sashimi-data" class="anchor"></a>Plot Sashimi Data</h3>
<p>Once the data is prepared, the plot function can be called and customized to fit the particular visualization.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ggGria1 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/plotSashimi.html">plotSashimi</a></span>(shGria1,
   <span class="dt">junc_color=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/jamba/topics/alpha2col">alpha2col</a></span>(<span class="st">"goldenrod1"</span>, <span class="fl">0.4</span>),
   <span class="dt">junc_fill=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/jamba/topics/alpha2col">alpha2col</a></span>(<span class="st">"goldenrod1"</span>, <span class="fl">0.4</span>),
   <span class="dt">show=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"coverage"</span>, <span class="st">"junction"</span>, <span class="st">"junctionLabels"</span>),
   <span class="dt">fill_scheme=</span><span class="st">"sample_id"</span>);
<span class="co">#&gt; Warning: Ignoring unknown aesthetics: fill</span>
<span class="co">#&gt; Scale for 'fill' is already present. Adding another scale for 'fill',</span>
<span class="co">#&gt; which will replace the existing scale.</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(ggGria1);</code></pre></div>
<p><img src="create-a-sashimi-plot_files/figure-html/plot_sashimi-1.png" width="768"></p>
<div id="adding-gene-exon-model" class="section level4">
<h4 class="hasAnchor">
<a href="#adding-gene-exon-model" class="anchor"></a>Adding gene-exon model</h4>
<p>To add the gene-exon model, currently the best method is to use something like the <code>cowplot</code> package to assemble multiple ggplot visualizations together. It also helps in this context to force the <code>facet_wrap</code> to have one column, using the <code>ncol=1</code> argument.</p>
<p>An example is shown below.</p>
<p>Note for the Sashimi plots, the x-axis labels are hidden, since the gene-exon plot shares the same range and values.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="cf">if</span> (<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">require</a></span>(cowplot)) {
   cpGria1 &lt;-<span class="st"> </span>cowplot<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/cowplot/topics/plot_grid">plot_grid</a></span>(
      ggGria1 <span class="op">+</span><span class="st"> </span><span class="kw">theme</span>(<span class="dt">axis.text.x=</span><span class="kw">element_blank</span>()) <span class="op">+</span><span class="st"> </span><span class="kw">xlab</span>(<span class="ot">NULL</span>),
      ggGria1exons <span class="op">+</span><span class="st"> </span><span class="kw">ggtitle</span>(<span class="ot">NULL</span>) <span class="op">+</span><span class="st"> </span><span class="kw">expand_limits</span>(<span class="dt">y=</span><span class="op">-</span><span class="dv">2</span>),
      <span class="dt">ncol=</span><span class="dv">1</span>,
      <span class="dt">align=</span><span class="st">"v"</span>, 
      <span class="dt">axis=</span><span class="st">"lr"</span>,
      <span class="dt">rel_heights=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">4</span>,<span class="dv">4</span>));
   <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(cpGria1);
}
<span class="co">#&gt; Loading required package: cowplot</span>
<span class="co">#&gt; Warning: package 'cowplot' was built under R version 3.5.2</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'cowplot'</span>
<span class="co">#&gt; The following object is masked from 'package:ggplot2':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     ggsave</span></code></pre></div>
<p><img src="create-a-sashimi-plot_files/figure-html/plot_sashimi_exons-1.png" width="768"></p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#overview">Overview</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by James M. Ward.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
