<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Add gaps between GRangesList regions — addGRLgaps • splicejam</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Add gaps between GRangesList regions — addGRLgaps" />

<meta property="og:description" content="Add gaps between GRangesList regions" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">splicejam</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.49.900</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/basic-rnaseq-workflow.html">Splicejam basic RNAseq workflow</a>
    </li>
    <li>
      <a href="../articles/create-a-sashimi-plot.html">Create a Sashimi Plot</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jmw86069/splicejam">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Add gaps between GRangesList regions</h1>
    <small class="dont-index">Source: <a href='https://github.com/jmw86069/splicejam/blob/master/R/jambio-tx.R'><code>R/jambio-tx.R</code></a></small>
    <div class="hidden name"><code>addGRLgaps.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Add gaps between GRangesList regions</p>
    
    </div>

    <pre class="usage"><span class='fu'>addGRLgaps</span>(<span class='no'>grl</span>, <span class='kw'>strandSpecific</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>gapname</span> <span class='kw'>=</span> <span class='st'>"gap"</span>,
  <span class='kw'>suffix</span> <span class='kw'>=</span> <span class='st'>"_v"</span>, <span class='kw'>newValues</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/list'>list</a></span>(<span class='kw'>feature_type</span> <span class='kw'>=</span> <span class='st'>"gap"</span>),
  <span class='kw'>default_feature_type</span> <span class='kw'>=</span> <span class='st'>"exon"</span>, <span class='kw'>feature_type_colname</span> <span class='kw'>=</span> <span class='st'>"feature_type"</span>,
  <span class='kw'>doSort</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>grl</th>
      <td><p>GRangesList object</p></td>
    </tr>
    <tr>
      <th>strandSpecific</th>
      <td><p>logical indicating whether the gaps are calculated
per strand, see <code><a href='getGRLgaps.html'>getGRLgaps()</a></code>.</p></td>
    </tr>
    <tr>
      <th>gapname, suffix</th>
      <td><p>character vector supplying the name to assign to new
gap GRanges elements, using <code><a href='https://www.rdocumentation.org/packages/jamba/topics/makeNames'>jamba::makeNames()</a></code> with <code>suffix</code> as
described to define non-duplicated names. If <code>gapname is NULL</code> then
no names are assigned to new gap GRanges entries, however when the
input <code>gr</code> GRanges object has names, the concatenation of gaps
causes names <code>""</code> to be assigned to all gap GRanges elements, which
are duplicated for multiple gaps.</p></td>
    </tr>
    <tr>
      <th>newValues</th>
      <td><p>list of values to add to the resulting gap GRanges,
whose names become <code><a href='https://www.rdocumentation.org/packages/base/topics/colnames'>colnames(grl@unlistData)</a></code>, and whose values are used
to populate each column. By default a colname <code>"feature_type"</code> is
added, with value <code>"gap"</code> added to each row. When <code>newValues</code> is <code>NULL</code>
then no values are added to the gaps GRanges. For every entry in
<code><a href='https://www.rdocumentation.org/packages/base/topics/names'>names(newValues)</a></code>, if the colname does not exist, it is created
and populated with <code>default_feature_type</code> as a default value,
prior to adding gaps.</p></td>
    </tr>
    <tr>
      <th>default_feature_type, feature_type_colname</th>
      <td><p>character values,
indicating the type and colname to populate in <code>values(grl@unlistData)</code>
prior to adding gaps. When <code>feature_type_colname</code> is NULL,
no action is taken. When <code>feature_type_colname</code> does not
exist in <code><a href='https://www.rdocumentation.org/packages/base/topics/colnames'>colnames(values(grl@unlistData))</a></code>, it is created
with values in <code>default_feature_type</code>. Also any colname
defined by <code>newValues</code> that does not already exist is also
created and populated with <code>default_feature_type</code>.</p></td>
    </tr>
    <tr>
      <th>doSort</th>
      <td><p>logical indicating whether to sort the resulting
GRanges objects. When <code>doSort=FALSE</code> the gaps are added to the end
of each input <code>grl</code> GRanges object. Note that the GrangesList object
is not sorted, only the GRanges objects within the GRangesList
are sorted.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>logical indicating whether to print verbose output.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>additional arguments are passed to <code><a href='getGRLgaps.html'>getGRLgaps()</a></code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>GRangesList object, sorted per GRangesList element
when <code>doSort=TRUE</code>. When <code>newValues</code>
is supplied, the values for gaps GRanges elements will be assigned,
otherwise any column values present in <code>gr</code> will be <code>NA</code> for
gaps elements. The names of gaps elements are assigned using
<code>gapname</code> then are made unique using <code><a href='https://www.rdocumentation.org/packages/jamba/topics/makeNames'>jamba::makeNames()</a></code>,
unless <code>gapname is NULL</code>.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function adds gaps between each GRanges region separately
for each GRangesList element, where
there is a gap between two GRanges for
the same seqnames. When <code>strandSpecific=TRUE</code> the gaps are
determined per strand.</p>
<p>This function is a wrapper around <code><a href='getGRLgaps.html'>getGRLgaps()</a></code>, which is then
concatenated to the input <code>gr</code> GRanges object using <code><a href='https://www.rdocumentation.org/packages/S4Vectors/topics/List-utils'>S4Vectors::pc()</a></code>.
When the input <code>grl</code> GRanges has column <code><a href='https://www.rdocumentation.org/packages/S4Vectors/topics/Vector-class'>S4Vectors::values()</a></code> then the
gaps GRanges object will have <code>NA</code> values used by default. To supply
values, use the <code>newValues</code> argument, which assigns name-value pairs.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>Other jam GRanges functions: <code><a href='addGRgaps.html'>addGRgaps</a></code>,
  <code><a href='annotateGRLfromGRL.html'>annotateGRLfromGRL</a></code>,
  <code><a href='annotateGRfromGR.html'>annotateGRfromGR</a></code>,
  <code><a href='assignGRLexonNames.html'>assignGRLexonNames</a></code>,
  <code><a href='closestExonToJunctions.html'>closestExonToJunctions</a></code>,
  <code><a href='combineGRcoverage.html'>combineGRcoverage</a></code>,
  <code><a href='exoncov2polygon.html'>exoncov2polygon</a></code>,
  <code><a href='findOverlapsGRL.html'>findOverlapsGRL</a></code>,
  <code><a href='getFirstStrandedFromGRL.html'>getFirstStrandedFromGRL</a></code>,
  <code><a href='getGRLgaps.html'>getGRLgaps</a></code>,
  <code><a href='getGRcoverageFromBw.html'>getGRcoverageFromBw</a></code>,
  <code><a href='getGRgaps.html'>getGRgaps</a></code>, <code><a href='grl2df.html'>grl2df</a></code>,
  <code><a href='make_ref2compressed.html'>make_ref2compressed</a></code>, <code><a href='sortGRL.html'>sortGRL</a></code>,
  <code><a href='spliceGR2junctionDF.html'>spliceGR2junctionDF</a></code>,
  <code><a href='stackJunctions.html'>stackJunctions</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>gr</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>GenomicRanges</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/GenomicRanges/topics/GRanges-class'>GRanges</a></span>(<span class='kw'>seqnames</span><span class='kw'>=</span><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/rep'>rep</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"chr1"</span>,<span class='st'>"chr2"</span>), <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='fl'>3</span>,<span class='fl'>2</span>)),
   <span class='kw'>ranges</span><span class='kw'>=</span><span class='kw pkg'>IRanges</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/IRanges/topics/IRanges-constructor'>IRanges</a></span>(<span class='kw'>start</span><span class='kw'>=</span><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='fl'>100</span>, <span class='fl'>300</span>, <span class='fl'>400</span>, <span class='fl'>300</span>, <span class='fl'>700</span>),
      <span class='kw'>end</span><span class='kw'>=</span><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='fl'>199</span>, <span class='fl'>450</span>, <span class='fl'>500</span>, <span class='fl'>600</span>, <span class='fl'>800</span>)),
   <span class='kw'>strand</span><span class='kw'>=</span><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/rep'>rep</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"+"</span>,<span class='st'>"-"</span>), <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='fl'>3</span>,<span class='fl'>2</span>)),
   <span class='kw'>feature_type</span><span class='kw'>=</span><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/rep'>rep</a></span>(<span class='st'>"exon"</span>, <span class='fl'>5</span>));
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/names'>names</a></span>(<span class='no'>gr</span>) <span class='kw'>&lt;-</span> <span class='kw pkg'>jamba</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/jamba/topics/makeNames'>makeNames</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/rep'>rep</a></span>(<span class='st'>"exon"</span>, <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/length'>length</a></span>(<span class='no'>gr</span>)));
<span class='no'>gr</span>;</div><div class='output co'>#&gt; GRanges object with 5 ranges and 1 metadata column:
#&gt;           seqnames    ranges strand | feature_type
#&gt;              &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;character&gt;
#&gt;   exon_v1     chr1   100-199      + |         exon
#&gt;   exon_v2     chr1   300-450      + |         exon
#&gt;   exon_v3     chr1   400-500      + |         exon
#&gt;   exon_v4     chr2   300-600      - |         exon
#&gt;   exon_v5     chr2   700-800      - |         exon
#&gt;   -------
#&gt;   seqinfo: 2 sequences from an unspecified genome; no seqlengths</div><div class='input'><span class='fu'><a href='addGRgaps.html'>addGRgaps</a></span>(<span class='no'>gr</span>);</div><div class='output co'>#&gt; GRanges object with 7 ranges and 1 metadata column:
#&gt;           seqnames    ranges strand | feature_type
#&gt;              &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;character&gt;
#&gt;   exon_v1     chr1   100-199      + |         exon
#&gt;    gap_v1     chr1   200-299      + |          gap
#&gt;   exon_v2     chr1   300-450      + |         exon
#&gt;   exon_v3     chr1   400-500      + |         exon
#&gt;   exon_v4     chr2   300-600      - |         exon
#&gt;    gap_v2     chr2   601-699      - |          gap
#&gt;   exon_v5     chr2   700-800      - |         exon
#&gt;   -------
#&gt;   seqinfo: 2 sequences from an unspecified genome; no seqlengths</div><div class='input'>
<span class='no'>grl</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>GenomicRanges</span><span class='kw ns'>::</span><span class='fu'>split</span>(<span class='no'>gr</span>, <span class='kw pkg'>GenomicRanges</span><span class='kw ns'>::</span><span class='fu'>seqnames</span>(<span class='no'>gr</span>));
<span class='no'>grl</span>;</div><div class='output co'>#&gt; GRangesList object of length 2:
#&gt; $chr1 
#&gt; GRanges object with 3 ranges and 1 metadata column:
#&gt;           seqnames    ranges strand | feature_type
#&gt;              &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;character&gt;
#&gt;   exon_v1     chr1   100-199      + |         exon
#&gt;   exon_v2     chr1   300-450      + |         exon
#&gt;   exon_v3     chr1   400-500      + |         exon
#&gt; 
#&gt; $chr2 
#&gt; GRanges object with 2 ranges and 1 metadata column:
#&gt;           seqnames  ranges strand | feature_type
#&gt;   exon_v4     chr2 300-600      - |         exon
#&gt;   exon_v5     chr2 700-800      - |         exon
#&gt; 
#&gt; -------
#&gt; seqinfo: 2 sequences from an unspecified genome; no seqlengths</div><div class='input'><span class='fu'>addGRLgaps</span>(<span class='no'>grl</span>);</div><div class='output co'>#&gt; GRangesList object of length 2:
#&gt; $chr1 
#&gt; GRanges object with 4 ranges and 1 metadata column:
#&gt;           seqnames    ranges strand | feature_type
#&gt;              &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;character&gt;
#&gt;   exon_v1     chr1   100-199      + |         exon
#&gt;    gap_v1     chr1   200-299      + |          gap
#&gt;   exon_v2     chr1   300-450      + |         exon
#&gt;   exon_v3     chr1   400-500      + |         exon
#&gt; 
#&gt; $chr2 
#&gt; GRanges object with 3 ranges and 1 metadata column:
#&gt;           seqnames  ranges strand | feature_type
#&gt;   exon_v4     chr2 300-600      - |         exon
#&gt;    gap_v2     chr2 601-699      - |          gap
#&gt;   exon_v5     chr2 700-800      - |         exon
#&gt; 
#&gt; -------
#&gt; seqinfo: 2 sequences from an unspecified genome; no seqlengths</div><div class='input'><span class='fu'>addGRLgaps</span>(<span class='no'>grl</span>, <span class='kw'>strandSpecific</span><span class='kw'>=</span><span class='fl'>FALSE</span>);</div><div class='output co'>#&gt; GRangesList object of length 2:
#&gt; $chr1 
#&gt; GRanges object with 4 ranges and 1 metadata column:
#&gt;           seqnames    ranges strand | feature_type
#&gt;              &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;character&gt;
#&gt;   exon_v1     chr1   100-199      + |         exon
#&gt;   exon_v2     chr1   300-450      + |         exon
#&gt;   exon_v3     chr1   400-500      + |         exon
#&gt;    gap_v1     chr1   200-299      * |          gap
#&gt; 
#&gt; $chr2 
#&gt; GRanges object with 3 ranges and 1 metadata column:
#&gt;           seqnames  ranges strand | feature_type
#&gt;   exon_v4     chr2 300-600      - |         exon
#&gt;   exon_v5     chr2 700-800      - |         exon
#&gt;    gap_v2     chr2 601-699      * |          gap
#&gt; 
#&gt; -------
#&gt; seqinfo: 2 sequences from an unspecified genome; no seqlengths</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by James M. Ward, Shannon Farris.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

