<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Flatten exons by gene or transcript — flattenExonsBy • splicejam</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Flatten exons by gene or transcript — flattenExonsBy" />

<meta property="og:description" content="Flatten exons by gene or transcript" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">splicejam</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.51.900</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/basic-rnaseq-workflow.html">Splicejam basic RNAseq workflow</a>
    </li>
    <li>
      <a href="../articles/create-a-sashimi-plot.html">Create a Sashimi Plot</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jmw86069/splicejam">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Flatten exons by gene or transcript</h1>
    <small class="dont-index">Source: <a href='https://github.com/jmw86069/splicejam/blob/master/R/jambio-tx.R'><code>R/jambio-tx.R</code></a></small>
    <div class="hidden name"><code>flattenExonsBy.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Flatten exons by gene or transcript</p>
    
    </div>

    <pre class="usage"><span class='fu'>flattenExonsBy</span>(<span class='no'>exonsByTx</span>, <span class='no'>tx2geneDF</span>, <span class='kw'>by</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"gene"</span>, <span class='st'>"tx"</span>),
  <span class='kw'>detectedTx</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>genes</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>txColname</span> <span class='kw'>=</span> <span class='st'>"transcript_id"</span>,
  <span class='kw'>geneColname</span> <span class='kw'>=</span> <span class='st'>"gene_name"</span>, <span class='kw'>cdsByTx</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>cdsByGene</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>exonsByTx</th>
      <td><p>GRangesList named by transcript, containing one or
more GRanges representing exons. This data is often produced
from <code>TxDb</code> data using <code><a href='https://www.rdocumentation.org/packages/GenomicFeatures/topics/transcriptsBy'>GenomicFeatures::exonsBy(...,by="tx")</a></code>.</p></td>
    </tr>
    <tr>
      <th>tx2geneDF</th>
      <td><p>data.frame containing at least two columns with
transcript and gene annotation, whose colnames are defined by
arguments <code>txColname</code> and <code>geneColname</code> respectively. When
using a GTF file, <code><a href='makeTx2geneFromGtf.html'>makeTx2geneFromGtf()</a></code> can be used to
create a <code>tx2geneDF</code> in <code>data.frame</code> format.</p></td>
    </tr>
    <tr>
      <th>by</th>
      <td><p>character string to group exons, <code>"gene"</code> groups multiple
transcripts per gene, and <code>"tx"</code> groups exons per transcript.
Note that in both cases, it combines <code>exonsByTx</code> and <code>cdsByTx</code>
when <code>cdsByTx</code> is also supplied.</p></td>
    </tr>
    <tr>
      <th>detectedTx</th>
      <td><p>character vector of detected transcripts, used to
subset the overall transcripts prior to producing a flattened gene
exon model.</p></td>
    </tr>
    <tr>
      <th>genes</th>
      <td><p>optional character vector, representing a subset of
genes for which flattened exons will be prepared. This argument
is useful when focusing on only one or a subset of genes.</p></td>
    </tr>
    <tr>
      <th>txColname</th>
      <td><p>character string indicating a column from
<code><a href='https://www.rdocumentation.org/packages/base/topics/colnames'>colnames(tx2geneDF)</a></code> used to identify transcripts.</p></td>
    </tr>
    <tr>
      <th>geneColname</th>
      <td><p>character string indicating a column from
<code><a href='https://www.rdocumentation.org/packages/base/topics/colnames'>colnames(tx2geneDF)</a></code> used to identify gene name, or gene symbol.</p></td>
    </tr>
    <tr>
      <th>cdsByTx</th>
      <td><p><code>GRangesList</code> named by transcript, containing <code>GRanges</code>
exons that only include CDS regions. This data is often produced
from <code>TxDb</code> data using <code><a href='https://www.rdocumentation.org/packages/GenomicFeatures/topics/transcriptsBy'>GenomicFeatures::cdsBy(...,by="tx")</a></code>.</p></td>
    </tr>
    <tr>
      <th>cdsByGene</th>
      <td><p><code>GRangesList</code> named by gene, containing <code>GRanges</code>
exons that only include CDS regions. This data is often produced
from <code>TxDb</code> data using <code><a href='https://www.rdocumentation.org/packages/GenomicFeatures/topics/transcriptsBy'>GenomicFeatures::cdsBy(...,by="gene")</a></code>.
Note this input is only used when <code>by="gene"</code>.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>logical indicating whether to print verbose output.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>GRangesList named by gene when <code>by="gene"</code> or transcript when
<code>by="tx"</code>, containing non-overlapping GRanges with exon names
as described above.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function takes as input:</p><ul>
<li><p><code>exonsByTx</code> as a <code>GRangesList</code> object
of transcript exons named by the <code>transcript_id</code>,</p></li>
<li><p><code>tx2geneDF</code> a <code>data.frame</code> with transcript-gene cross-reference,</p></li>
<li><p><code>detectedTx</code> an optional character vector of <code>transcript_id</code>
values, used to subset the overall transcripts</p></li>
<li><p><code>cdsByTx</code> an optional <code>GRangesList</code> object, similar to <code>exonsByTx</code>
except that it only contains the CDS portion of exons</p></li>
</ul>
    <p>This function groups exons together by gene, producing a flattened,
disjoint (non-overlapping) set of exons, where exons are subdivided
when there are multiple boundaries.</p>
<p>Finally, it labels each exon using a defined naming scheme:</p><ul>
<li><p>Each contiguous exon is numbered in order, starting at <code>1</code> for the
first stranded exon for the gene, for example <code>exon1</code>, <code>exon2</code>,
<code>exon3</code>.</p></li>
<li><p>When an exon is sub-divided, each section is labeled with
an alphabetic suffix to indicate the order within that exon,
for example <code>exon1a</code>, <code>exon1b</code>, <code>exon1c</code>.</p></li>
</ul>
    <p>A text schematic is shown below:</p>
<p><code>|=======|======|......|=======|.....|=======|=======|=======|</code></p>
<p><code>|.exon1a|exon1b|......|.exon2.|.....|.exon3a|.exon3b|.exon3c|</code></p>
<p>Where</p><ul>
<li><p><code>|====|</code> represents an exon,</p></li>
<li><p><code>|====|====|</code> represents one contiguous exon with two
sub-divided parts, and</p></li>
<li><p><code>.....</code> represents an intron.</p></li>
</ul>
    <p>It is recommended but not required to supply <code>detectedTx</code>,
since it can greatly reduce the total number of transcripts.
This step has two benefits: It can greatly simplify the
resulting exon model based upon observed data, and it
has the by-product of removing potentially erroneous
transcripts which often has no supporting observed data.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>Other jam RNA-seq functions: <code><a href='assignGRLexonNames.html'>assignGRLexonNames</a></code>,
  <code><a href='closestExonToJunctions.html'>closestExonToJunctions</a></code>,
  <code><a href='combineGRcoverage.html'>combineGRcoverage</a></code>,
  <code><a href='defineDetectedTx.html'>defineDetectedTx</a></code>,
  <code><a href='detectedTxInfo.html'>detectedTxInfo</a></code>,
  <code><a href='exoncov2polygon.html'>exoncov2polygon</a></code>,
  <code><a href='groups2contrasts.html'>groups2contrasts</a></code>,
  <code><a href='makeTx2geneFromGtf.html'>makeTx2geneFromGtf</a></code>,
  <code><a href='make_ref2compressed.html'>make_ref2compressed</a></code>,
  <code><a href='runDiffSplice.html'>runDiffSplice</a></code>, <code><a href='sortSamples.html'>sortSamples</a></code>,
  <code><a href='spliceGR2junctionDF.html'>spliceGR2junctionDF</a></code></p></div>
    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#see-also">See also</a></li>
          </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by James M. Ward, Shannon Farris.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

