<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Create a ref2compressed function to compress GR gaps — make_ref2compressed • splicejam</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Create a ref2compressed function to compress GR gaps — make_ref2compressed" />

<meta property="og:description" content="Create a ref2compressed function to compress GR gaps" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">splicejam</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.30.900</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/basic-rnaseq-workflow.html">Splicejam basic RNAseq workflow</a>
    </li>
    <li>
      <a href="../articles/create-a-sashimi-plot.html">Create a Sashimi Plot</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jmw86069/splicejam">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Create a ref2compressed function to compress GR gaps</h1>
    <small class="dont-index">Source: <a href='https://github.com/jmw86069/splicejam/blob/master/R/jambio-plot.R'><code>R/jambio-plot.R</code></a></small>
    <div class="hidden name"><code>make_ref2compressed.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Create a ref2compressed function to compress GR gaps</p>
    
    </div>

    <pre class="usage"><span class='fu'>make_ref2compressed</span>(<span class='no'>gr</span>, <span class='kw'>gapWidth</span> <span class='kw'>=</span> <span class='fl'>200</span>, <span class='kw'>keepValues</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>upstream</span> <span class='kw'>=</span> <span class='fl'>50000</span>, <span class='kw'>upstreamGapWidth</span> <span class='kw'>=</span> <span class='no'>gapWidth</span> * <span class='fl'>3</span>,
  <span class='kw'>downstream</span> <span class='kw'>=</span> <span class='fl'>50000</span>, <span class='kw'>downstreamGapWidth</span> <span class='kw'>=</span> <span class='no'>gapWidth</span> * <span class='fl'>3</span>, <span class='kw'>nBreaks</span> <span class='kw'>=</span> <span class='fl'>7</span>,
  <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>gr</th>
      <td><p>GRanges object containing regions not to compress. Regions
which are unstranded gaps are compressed to fixed width.</p></td>
    </tr>
    <tr>
      <th>gapWidth</th>
      <td><p>integer value used for fixed gap width, or when
NULL the gap width is defined as 3 times the median feature width.</p></td>
    </tr>
    <tr>
      <th>keepValues</th>
      <td><p>logical indicating whether to keep feature values
in the GRanges data.</p></td>
    </tr>
    <tr>
      <th>upstream, downstream, upstreamGapWidth, downstreamGapWidth</th>
      <td><p>used
to define the compression of coordinates upstream and downstream
the supplied GRanges. In reality, the upstream range and upstream
gap width defines a multiplier, and all upstream coordinates are
compressed through zero. Similarly, all downstream coordinates
are compressed to 10 billion, which is roughly 3 times the size
of the human genome.</p></td>
    </tr>
    <tr>
      <th>nBreaks</th>
      <td><p>the default number of x-axis coordinate breaks used
in ggplot labeling.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>logical indicating whether to print verbose output.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>additional arguments are ignored.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>list with <code>trans_grc</code> which is class <code>"trans"</code> suitable
for use in ggplot2 functions; <code>transform</code> a function that converts
chromosome coordinates to compressed coordinates; <code>inverse</code> a function
that converts compressed coordinates to chromosome coordinates;
<code>scale_x_grc</code> a function used similar to <code><a href='https://www.rdocumentation.org/packages/ggplot2/topics/scale_continuous'>ggplot2::scale_x_continuous()</a></code>
during ggplot2 creation; <code>gr</code> a function that compresses coordinates
in a <code>GRanges</code> object; <code>grl</code> a function that compresses coordinates in
a <code>GRangesList</code> object. Attributes <code>"lookupCoordDF"</code> is a two-column
data.frame with chromosome coordinates and compressed coordinates,
which is used to create the other transformation functions via
<code><a href='https://www.rdocumentation.org/packages/stats/topics/approxfun'>stats::approx()</a></code>; <code>"gapWidth"</code> the gap width used, since it can
be programmatically defined; <code>"gr"</code> the <code>GRanges</code> input data used
to train the transformation.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function takes a set of GRanges which are to be maintained
with fixed aspect ratio, and it defines a function to compress
coordinates of the gaps between GRanges features.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='grl2df.html'>grl2df()</a></code>, <code>test_junc_wide_gr</code></p>
<p>Other jam GRanges functions: <code><a href='addGRLgaps.html'>addGRLgaps</a></code>,
  <code><a href='addGRgaps.html'>addGRgaps</a></code>,
  <code><a href='annotateGRLfromGRL.html'>annotateGRLfromGRL</a></code>,
  <code><a href='annotateGRfromGR.html'>annotateGRfromGR</a></code>,
  <code><a href='assignGRLexonNames.html'>assignGRLexonNames</a></code>,
  <code><a href='closestExonToJunctions.html'>closestExonToJunctions</a></code>,
  <code><a href='combineGRcoverage.html'>combineGRcoverage</a></code>,
  <code><a href='exoncov2polygon.html'>exoncov2polygon</a></code>,
  <code><a href='findOverlapsGRL.html'>findOverlapsGRL</a></code>,
  <code><a href='getFirstStrandedFromGRL.html'>getFirstStrandedFromGRL</a></code>,
  <code><a href='getGRLgaps.html'>getGRLgaps</a></code>,
  <code><a href='getGRcoverageFromBw.html'>getGRcoverageFromBw</a></code>,
  <code><a href='getGRgaps.html'>getGRgaps</a></code>, <code><a href='grl2df.html'>grl2df</a></code>,
  <code><a href='sortGRL.html'>sortGRL</a></code>, <code><a href='spliceGR2junctionDF.html'>spliceGR2junctionDF</a></code>,
  <code><a href='stackJunctions.html'>stackJunctions</a></code></p>
<p>Other splicejam core functions: <code><a href='exoncov2polygon.html'>exoncov2polygon</a></code>,
  <code><a href='gene2gg.html'>gene2gg</a></code>, <code><a href='grl2df.html'>grl2df</a></code>,
  <code><a href='plotSashimi.html'>plotSashimi</a></code>, <code><a href='prepareSashimi.html'>prepareSashimi</a></code></p>
<p>Other jam RNA-seq functions: <code><a href='assignGRLexonNames.html'>assignGRLexonNames</a></code>,
  <code><a href='closestExonToJunctions.html'>closestExonToJunctions</a></code>,
  <code><a href='combineGRcoverage.html'>combineGRcoverage</a></code>,
  <code><a href='defineDetectedTx.html'>defineDetectedTx</a></code>,
  <code><a href='detectedTxInfo.html'>detectedTxInfo</a></code>,
  <code><a href='exoncov2polygon.html'>exoncov2polygon</a></code>,
  <code><a href='flattenExonsBy.html'>flattenExonsBy</a></code>,
  <code><a href='groups2contrasts.html'>groups2contrasts</a></code>,
  <code><a href='makeTx2geneFromGtf.html'>makeTx2geneFromGtf</a></code>,
  <code><a href='runDiffSplice.html'>runDiffSplice</a></code>, <code><a href='sortSamples.html'>sortSamples</a></code>,
  <code><a href='spliceGR2junctionDF.html'>spliceGR2junctionDF</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/message'>suppressPackageStartupMessages</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/library'>library</a></span>(<span class='no'>GenomicRanges</span>));
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/message'>suppressPackageStartupMessages</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/library'>library</a></span>(<span class='no'>ggplot2</span>));

<span class='fu'><a href='https://www.rdocumentation.org/packages/utils/topics/data'>data</a></span>(<span class='no'>test_exon_wide_gr</span>);
<span class='co'># To plot a simple GRanges object</span>
<span class='no'>widedf</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='grl2df.html'>grl2df</a></span>(<span class='no'>test_exon_wide_gr</span>);
<span class='no'>ggWide</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/ggplot2/topics/ggplot'>ggplot</a></span>(<span class='no'>widedf</span>, <span class='fu'><a href='https://www.rdocumentation.org/packages/ggplot2/topics/aes'>aes</a></span>(<span class='kw'>x</span><span class='kw'>=</span><span class='no'>x</span>, <span class='kw'>y</span><span class='kw'>=</span><span class='no'>y</span>, <span class='kw'>group</span><span class='kw'>=</span><span class='no'>id</span>, <span class='kw'>fill</span><span class='kw'>=</span><span class='no'>feature_type</span>)) +
   <span class='fu'><a href='https://www.rdocumentation.org/packages/ggplot2/topics/geom_polygon'>geom_polygon</a></span>() +
   <span class='kw pkg'>colorjam</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/colorjam/topics/theme_jam'>theme_jam</a></span>() +
   <span class='kw pkg'>colorjam</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/colorjam/topics/scale_fill_jam'>scale_fill_jam</a></span>() +
   <span class='fu'><a href='https://www.rdocumentation.org/packages/ggplot2/topics/labs'>xlab</a></span>(<span class='st'>"chr1"</span>) +
   <span class='fu'><a href='https://www.rdocumentation.org/packages/ggplot2/topics/labs'>ggtitle</a></span>(<span class='st'>"exons (introns as-is)"</span>)
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/print'>print</a></span>(<span class='no'>ggWide</span>);</div><div class='img'><img src='make_ref2compressed-1.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'># Now compress the introns keeping axis labels</span>
<span class='no'>ref2c</span> <span class='kw'>&lt;-</span> <span class='fu'>make_ref2compressed</span>(<span class='no'>test_exon_wide_gr</span>,
   <span class='kw'>nBreaks</span><span class='kw'>=</span><span class='fl'>10</span>);
<span class='no'>ggWide2</span> <span class='kw'>&lt;-</span> <span class='no'>ggWide</span> +
   <span class='fu'><a href='https://www.rdocumentation.org/packages/ggplot2/topics/scale_continuous'>scale_x_continuous</a></span>(<span class='kw'>trans</span><span class='kw'>=</span><span class='no'>ref2c</span>$<span class='no'>trans_grc</span>) +
   <span class='fu'><a href='https://www.rdocumentation.org/packages/ggplot2/topics/labs'>xlab</a></span>(<span class='st'>"chr1 (compressed introns)"</span>) +
   <span class='fu'><a href='https://www.rdocumentation.org/packages/ggplot2/topics/labs'>ggtitle</a></span>(<span class='st'>"exons (compressed introns)"</span>)
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/print'>print</a></span>(<span class='no'>ggWide2</span>);</div><div class='img'><img src='make_ref2compressed-2.png' alt='' width='700' height='433' /></div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by James M. Ward.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

