<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Prepare Sashimi plot data — prepareSashimi • splicejam</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Prepare Sashimi plot data — prepareSashimi" />

<meta property="og:description" content="Prepare Sashimi plot data" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">splicejam</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.30.900</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/basic-rnaseq-workflow.html">Splicejam basic RNAseq workflow</a>
    </li>
    <li>
      <a href="../articles/create-a-sashimi-plot.html">Create a Sashimi Plot</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jmw86069/splicejam">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Prepare Sashimi plot data</h1>
    <small class="dont-index">Source: <a href='https://github.com/jmw86069/splicejam/blob/master/R/jambio-plot.R'><code>R/jambio-plot.R</code></a></small>
    <div class="hidden name"><code>prepareSashimi.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Prepare Sashimi plot data</p>
    
    </div>

    <pre class="usage"><span class='fu'>prepareSashimi</span>(<span class='kw'>flatExonsByGene</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>filesDF</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='no'>gene</span>,
  <span class='kw'>sample_id</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>minJunctionScore</span> <span class='kw'>=</span> <span class='fl'>10</span>, <span class='kw'>gapWidth</span> <span class='kw'>=</span> <span class='fl'>200</span>,
  <span class='kw'>addGaps</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>baseline</span> <span class='kw'>=</span> <span class='fl'>0</span>, <span class='kw'>compressGR</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>ref2c</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>gap_feature_type</span> <span class='kw'>=</span> <span class='st'>"intron"</span>, <span class='kw'>default_feature_type</span> <span class='kw'>=</span> <span class='st'>"exon"</span>,
  <span class='kw'>feature_type_colname</span> <span class='kw'>=</span> <span class='st'>"feature_type"</span>, <span class='kw'>exon_label_type</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"none"</span>,
  <span class='st'>"repel"</span>, <span class='st'>"mark"</span>), <span class='kw'>junc_label_type</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"repel"</span>, <span class='st'>"mark"</span>, <span class='st'>"none"</span>),
  <span class='kw'>return_data</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"ggCov"</span>, <span class='st'>"ggJunc"</span>, <span class='st'>"ggSashimi"</span>, <span class='st'>"covDF"</span>, <span class='st'>"juncDF"</span>,
  <span class='st'>"ref2c"</span>, <span class='st'>"all"</span>), <span class='kw'>junc_color</span> <span class='kw'>=</span> <span class='fu'>alpha2col</span>(<span class='st'>"goldenrod3"</span>, <span class='fl'>0.7</span>),
  <span class='kw'>junc_fill</span> <span class='kw'>=</span> <span class='fu'>alpha2col</span>(<span class='st'>"goldenrod1"</span>, <span class='fl'>0.4</span>), <span class='kw'>doStackJunctions</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>coord_method</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"coord"</span>, <span class='st'>"scale"</span>, <span class='st'>"none"</span>), <span class='kw'>scoreFactor</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>scoreArcFactor</span> <span class='kw'>=</span> <span class='fl'>0.2</span>, <span class='kw'>scoreArcMinimum</span> <span class='kw'>=</span> <span class='fl'>100</span>, <span class='kw'>covGR</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>juncGR</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>do_shiny_progress</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>flatExonsByGene</th>
      <td><p>GRangesList named by gene, whose GRanges
elements are flattened, disjoint, non-overlapping genomic ranges
per gene.</p></td>
    </tr>
    <tr>
      <th>filesDF</th>
      <td><p>data.frame with columns <code>url</code>, <code>sample_id</code>, <code>type</code>,
where: <code>url</code> is any valid file path or URL compatible with
<code>base::read.table()</code>; <code>sample_id</code> is an identified representing
a biological sample, used to group common files together;
<code>type</code> is one of <code>"bw"</code> for bigWig coverage, <code>"junction"</code> for
BED12 format splice junctions.</p></td>
    </tr>
    <tr>
      <th>gene</th>
      <td><p>character string of the gene to prepare, which must be
present in <code><a href='https://www.rdocumentation.org/packages/base/topics/names'>names(flatExonsByGene)</a></code>.</p></td>
    </tr>
    <tr>
      <th>gapWidth</th>
      <td><p>numeric value of the fixed width to use for
gaps (introns) between exon features. If <code>NULL</code> then
<code><a href='getGRgaps.html'>getGRgaps()</a></code> will use the default based upon the median exon
width.</p></td>
    </tr>
    <tr>
      <th>addGaps</th>
      <td><p>logical indicating whether to include gap regions
in the coverage plot, for example including introns or intergenic
regions. When <code>compressGR=TRUE</code> then gaps regions are
down-sampled using running maximum signal with roughly the same
x-axis resolution as uncompressed regions.</p></td>
    </tr>
    <tr>
      <th>baseline</th>
      <td><p>numeric vector named by <code><a href='https://www.rdocumentation.org/packages/base/topics/names'>names(flatExonsByGene)</a></code>
where baseline is used to adjust the y-axis baseline position
above or below zero.</p></td>
    </tr>
    <tr>
      <th>compressGR</th>
      <td><p>logical indicating whether to compress GRanges
coordinates in the output data, where gaps/introns are set
to a fixed width. When <code>ref2c</code> is not supplied, and
<code>compressGR=TRUE</code>, then <code>ref2c</code> is created using
<code><a href='make_ref2compressed.html'>make_ref2compressed()</a></code>.</p></td>
    </tr>
    <tr>
      <th>ref2c</th>
      <td><p>list object output from <code><a href='make_ref2compressed.html'>make_ref2compressed()</a></code> used
to compress axis coordinates, to compress polygon coverage
data in compressed regions, and to adjust splice junction arcs
using compressed coordinates.</p></td>
    </tr>
    <tr>
      <th>gap_feature_type</th>
      <td><p>the default feature_type value to use for
gaps when <code>addGaps=TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>doStackJunctions</th>
      <td><p>logical indicating whether to stack
junction arcs at each end, this argument is passed to
<code><a href='grl2df.html'>grl2df()</a></code> which calls <code><a href='stackJunctions.html'>stackJunctions()</a></code>.</p></td>
    </tr>
    <tr>
      <th>covGR</th>
      <td><p>GRanges object containing coverage data in columns
stored as NumericList class, where <code><a href='https://www.rdocumentation.org/packages/base/topics/colnames'>colnames(values(covGR))</a></code>
are present in <code>filesDF$url</code> when <code>files$type %in% "coverage_gr"</code>.</p></td>
    </tr>
    <tr>
      <th>juncGR</th>
      <td><p>GRanges object containing splice junctions, where
<code>"score"</code> is used for the abundance of splice junction reads,
and <code>"sample_id"</code> is used to define the biological <code>sample_id</code>.</p></td>
    </tr>
    <tr>
      <th>do_shiny_progress</th>
      <td><p>logical indicating whether to send
progress updates to a running shiny app, using the
<code><a href='https://www.rdocumentation.org/packages/shiny/topics/withProgress'>shiny::withProgress()</a></code> and <code><a href='https://www.rdocumentation.org/packages/shiny/topics/withProgress'>shiny::incProgress()</a></code> methods.
This function only calls <code><a href='https://www.rdocumentation.org/packages/shiny/topics/withProgress'>shiny::incProgress()</a></code> and
assumes the <code><a href='https://www.rdocumentation.org/packages/shiny/topics/withProgress'>shiny::withProgress()</a></code> has already been
initialized.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>logical indicating whether to print verbose output.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>additional arguments are passed to <code><a href='make_ref2compressed.html'>make_ref2compressed()</a></code>,
<code><a href='getGRcoverageFromBw.html'>getGRcoverageFromBw()</a></code>, <code><a href='exoncov2polygon.html'>exoncov2polygon()</a></code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>list containing <code>ggSashimi</code> a ggplot2 graphical object
containing a full Sashimi plot; <code>ggCov</code> the RNA-seq coverage
subset of the Sashimi plot; <code>ggJunc</code> the splice junction
subsset of the Sashimi plot; <code>ref2c</code> the output of
<code><a href='make_ref2compressed.html'>make_ref2compressed()</a></code> used for ggplot2 coordinate
visualization; <code>covDF</code>, <code>juncDF</code> data.frame objects
with the raw data used to create ggplot2 objects;
<code>covGR</code>, <code>juncGR</code> the GRanges objects used to create
the data.frames; <code>gr</code> the GRanges object representing the
exons for the gene of interest; <code>juncLabelDF</code> the data.frame
containing exon label coordinates used to add labels to
the splice junction arcs.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function is the workhorse function used to produce
Sashimi plots, and is intended to be a convenient wrapper
function for several other individual functions.</p>
<p>At a minimum, a Sashimi plot requires three things:</p><ol>
<li><p>Exons, usually from a gene of interest.</p></li>
<li><p>RNA-seq coverage data.</p></li>
<li><p>Splice junction data.</p></li>
</ol>
    <p>There is some required pre-processing before running
<code>prepareSashimi()</code>:</p><ul>
<li><p>Prepare flattened exons by gene using <code>flattenExonsByGene()</code>
and corresponding data, including <code>exonsByGene</code>, <code>cdsByGene</code>,
and <code>tx2geneDF</code>. Verify the gene exon model data using
<code><a href='gene2gg.html'>gene2gg()</a></code>.</p></li>
<li><p>Find file paths, or web URLs, for a set of bigWig coverage
files, representing RNA-seq coverage for each strand, for
the samples of interest. Test the coverage data using
<code><a href='getGRcoverageFromBw.html'>getGRcoverageFromBw()</a></code> for a small set of GRanges data.</p></li>
<li><p>Find file paths, or web URLs, for a set of BED6 or BED12
format files, note that it cannot currently use bigBed format
due to limitations in the <code>rtracklayer</code> package.
Test the splice junction data using <code><a href='https://www.rdocumentation.org/packages/rtracklayer/topics/export'>rtracklayer::import()</a></code>
for a small range of GRanges features, then send the data
to <code><a href='spliceGR2junctionDF.html'>spliceGR2junctionDF()</a></code> to prepare a data.frame summary.</p></li>
</ul>
    <p>The basic input for coverage and junction data is a data.frame,
which defines each file path or url, the type of data
<code>"bw"</code> or <code>"junction"</code>, and the biological sample <code>"sample_id"</code>.
Any file path compatible with <code><a href='https://www.rdocumentation.org/packages/rtracklayer/topics/export'>rtracklayer::import()</a></code> will
work, including web URLs and local files. When using a web URL
you may need to use <code>"https://"</code> format to force the use
of secure web requests, but this requirement varies by country.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>Other RNA-seq functions: <code><a href='getGRcoverageFromBw.html'>getGRcoverageFromBw</a></code></p>
<p>Other jam plot functions: <code><a href='bgaPlotly3d.html'>bgaPlotly3d</a></code>,
  <code><a href='factor2label.html'>factor2label</a></code>, <code><a href='gene2gg.html'>gene2gg</a></code>,
  <code><a href='grl2df.html'>grl2df</a></code>, <code><a href='plotSashimi.html'>plotSashimi</a></code>,
  <code><a href='stackJunctions.html'>stackJunctions</a></code></p>
<p>Other splicejam core functions: <code><a href='exoncov2polygon.html'>exoncov2polygon</a></code>,
  <code><a href='gene2gg.html'>gene2gg</a></code>, <code><a href='grl2df.html'>grl2df</a></code>,
  <code><a href='make_ref2compressed.html'>make_ref2compressed</a></code>,
  <code><a href='plotSashimi.html'>plotSashimi</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># The active example below uses sample data</span>
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/message'>suppressPackageStartupMessages</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/library'>library</a></span>(<span class='no'>GenomicRanges</span>));

<span class='fu'><a href='https://www.rdocumentation.org/packages/utils/topics/data'>data</a></span>(<span class='no'>test_exon_gr</span>);
<span class='fu'><a href='https://www.rdocumentation.org/packages/utils/topics/data'>data</a></span>(<span class='no'>test_junc_gr</span>);
<span class='fu'><a href='https://www.rdocumentation.org/packages/utils/topics/data'>data</a></span>(<span class='no'>test_cov_gr</span>);
<span class='no'>filesDF</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/data.frame'>data.frame</a></span>(<span class='kw'>url</span><span class='kw'>=</span><span class='st'>"sample_A"</span>,
   <span class='kw'>type</span><span class='kw'>=</span><span class='st'>"coverage_gr"</span>,
   <span class='kw'>sample_id</span><span class='kw'>=</span><span class='st'>"sample_A"</span>);
<span class='no'>sh1</span> <span class='kw'>&lt;-</span> <span class='fu'>prepareSashimi</span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/GenomicRanges/topics/GRangesList-class'>GRangesList</a></span>(<span class='kw'>TestGene1</span><span class='kw'>=</span><span class='no'>test_exon_gr</span>),
   <span class='kw'>filesDF</span><span class='kw'>=</span><span class='no'>filesDF</span>,
   <span class='kw'>gene</span><span class='kw'>=</span><span class='st'>"TestGene1"</span>,
   <span class='kw'>covGR</span><span class='kw'>=</span><span class='no'>test_cov_gr</span>,
   <span class='kw'>juncGR</span><span class='kw'>=</span><span class='no'>test_junc_gr</span>);
<span class='fu'><a href='plotSashimi.html'>plotSashimi</a></span>(<span class='no'>sh1</span>);</div><div class='output co'>#&gt; <span class='message'>Scale for 'fill' is already present. Adding another scale for 'fill', which</span>
#&gt; <span class='message'>will replace the existing scale.</span></div><div class='img'><img src='prepareSashimi-1.png' alt='' width='700' height='433' /></div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by James M. Ward.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

