<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Perform differential isoform analysis using diffSplice — runDiffSplice • splicejam</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Perform differential isoform analysis using diffSplice — runDiffSplice" />

<meta property="og:description" content="Perform differential isoform analysis using diffSplice" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">splicejam</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.14.900</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/basic-rnaseq-workflow.html">Splicejam basic RNAseq workflow</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jmw86069/splicejam">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Perform differential isoform analysis using diffSplice</h1>
    <small class="dont-index">Source: <a href='https://github.com/jmw86069/splicejam/blob/master/R/jambio-tx.R'><code>R/jambio-tx.R</code></a></small>
    <div class="hidden name"><code>runDiffSplice.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Perform differential isoform analysis using diffSplice</p>
    
    </div>

    <pre class="usage"><span class='fu'>runDiffSplice</span>(<span class='no'>iMatrixTx</span>, <span class='kw'>detectedTx</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/colnames'>rownames</a></span>(<span class='no'>iMatrixTx</span>), <span class='no'>tx2geneDF</span>,
  <span class='kw'>txColname</span> <span class='kw'>=</span> <span class='st'>"transcript_id"</span>, <span class='kw'>geneColname</span> <span class='kw'>=</span> <span class='st'>"gene_name"</span>, <span class='no'>iDesign</span>,
  <span class='no'>iContrasts</span>, <span class='kw'>cutoffFDR</span> <span class='kw'>=</span> <span class='fl'>0.05</span>, <span class='kw'>cutoffFold</span> <span class='kw'>=</span> <span class='fl'>1.5</span>,
  <span class='kw'>collapseByGene</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>spliceTest</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"t"</span>, <span class='st'>"simes"</span>, <span class='st'>"F"</span>),
  <span class='kw'>sep</span> <span class='kw'>=</span> <span class='st'>" "</span>, <span class='kw'>useVoom</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>iMatrixTx</th>
      <td><p>numeric matrix of expression, with transcripts as
rows and samples as columns. The data is assumed to be log2-transformed
using the format <code><a href='https://www.rdocumentation.org/packages/base/topics/Log'>log2(1 + x)</a></code>. This data should be normalized using
appropriate methods, outside the scope of this function.</p></td>
    </tr>
    <tr>
      <th>detectedTx</th>
      <td><p>character vector containing all or a subset of
<code><a href='https://www.rdocumentation.org/packages/base/topics/colnames'>rownames(iMatrix)</a></code> used for statistical testing.</p></td>
    </tr>
    <tr>
      <th>tx2geneDF</th>
      <td><p>data.frame with colnames <code><a href='https://www.rdocumentation.org/packages/base/topics/c'>c(txColname, geneColname)</a></code>,
where all entries of <code><a href='https://www.rdocumentation.org/packages/base/topics/colnames'>rownames(iMatrix)</a></code> are represented
in <code>tx2geneDF[,txColname]</code>.</p></td>
    </tr>
    <tr>
      <th>txColname, geneColname</th>
      <td><p>the <code><a href='https://www.rdocumentation.org/packages/base/topics/colnames'>colnames(tx2geneDF)</a></code> representing
the <code><a href='https://www.rdocumentation.org/packages/base/topics/colnames'>rownames(iMatrixTx)</a></code> matched by <code>tx2geneDF[,txColname]</code>,
and the associated genes given by <code>tx2geneDF[,geneColname]</code>.
Note that <code>detectedTx</code> must also contain values in <code><a href='https://www.rdocumentation.org/packages/base/topics/colnames'>rownames(iMatrixTx)</a></code>
and <code>tx2geneDF[,txColname]</code>.</p></td>
    </tr>
    <tr>
      <th>iDesign</th>
      <td><p>numeric matrix representing the design matrix for
the experiment design. For example, <code>limma::model.matrix(~0+group)</code>
will represent each group. Typically, <code><a href='https://www.rdocumentation.org/packages/base/topics/colnames'>rownames(iDesign)</a></code> should
be defined to match the <code><a href='https://www.rdocumentation.org/packages/base/topics/colnames'>colnames(iMatrix)</a></code> even if it requires
extra processing. The <code><a href='https://www.rdocumentation.org/packages/base/topics/colnames'>colnames(iDesign)</a></code> should represent
group names used in <code><a href='https://www.rdocumentation.org/packages/base/topics/colnames'>rownames(iContrasts)</a></code>.</p></td>
    </tr>
    <tr>
      <th>iContrasts</th>
      <td><p>numeric matrix representing the contrasts used
in statistical comparisons. This matrix can be generated by
running <code><a href='https://www.rdocumentation.org/packages/limma/topics/makeContrasts'>limma::makeContrasts()</a></code> using a format similar to the
following: <code><a href='https://www.rdocumentation.org/packages/limma/topics/makeContrasts'>limma::makeContrasts(contrasts="group1-group2", levels=iDesign)</a></code>,
see "Examples" for more info.</p></td>
    </tr>
    <tr>
      <th>cutoffFDR</th>
      <td><p>numeric value indicating a statistical threshold
on the FDR (adjusted P-value). Values should be between 0 and 1, where
<code>cutoffFDR=1</code> would impose no threshold on the adjusted P-value.</p></td>
    </tr>
    <tr>
      <th>cutoffFold</th>
      <td><p>numeric value indicating the minimum normal space
fold change allowed for statistical hits. For example <code>cutoffFold=2</code>
would require a 2-fold change, equivalent to log2 fold change &gt;= 1.</p></td>
    </tr>
    <tr>
      <th>collapseByGene</th>
      <td><p>logical indicating whether results should be
summarized at the gene level after filtering statistical hits.</p></td>
    </tr>
    <tr>
      <th>spliceTest</th>
      <td><p>character value described in <code><a href='https://www.rdocumentation.org/packages/limma/topics/topSplice'>limma::topSplice()</a></code>
which defines the statistical test to return. The default <code>"t"</code>
returns the t-test result for each isoform, which is mainly beneficial
because it also includes fold change that can be filtered. The
<code>"F"</code> returns F-test per gene, and <code>"simes"</code> returns the per-gene
t-test P-value after Simes adjustment per gene.</p></td>
    </tr>
    <tr>
      <th>sep</th>
      <td><p>character value used as a delimiter in output data.frame
colnames, such that each stats is followed by the contrast name,
separated by this delimiter.</p></td>
    </tr>
    <tr>
      <th>useVoom</th>
      <td><p>logical indicating whether to apply the <code><a href='https://www.rdocumentation.org/packages/limma/topics/voom'>limma::voom()</a></code>
adjustment prior to running <code><a href='https://www.rdocumentation.org/packages/limma/topics/diffSplice'>limma::diffSplice()</a></code>. This value
should be <code>TRUE</code> when analyzing count or pseudocount data.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>logical indicating whether to print verbose output.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>additional arguments are ignored.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>list containing <code>detectedTx</code>, <code>detectedTxUse</code> the detectedTx
values representing genes with multiple transcripts; <code>fit</code> the initial
model fit; <code>fit2</code> the contrast model fit; <code>splice</code> the output from
<code><a href='https://www.rdocumentation.org/packages/limma/topics/diffSplice'>limma::diffSplice()</a></code>; <code>statsDFs</code> list of data.frame output from
<code><a href='https://www.rdocumentation.org/packages/limma/topics/topSplice'>limma::topSplice()</a></code> either at the transcript level or the gene level.</p>
<p>Note that when <code>collapseByGene=TRUE</code> the results will return the first
transcript entry per gene that has the best P-value. Often one gene
will have two transcripts with identical P-value, and the order
that the transcripts appear is inconsistent. Therefore, the direction
of fold change is not meaningful by itself, except with respect to
the specific isoform returned.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function is intended to be a straightforward method
to call <code><a href='https://www.rdocumentation.org/packages/limma/topics/diffSplice'>limma::diffSplice()</a></code> and return the output of
<code><a href='https://www.rdocumentation.org/packages/limma/topics/topSplice'>limma::topSplice()</a></code> in helpful formats for downstream
use.</p>
<p>For example, it can be helpful to call this function with and without
the "voom" component, dependent upon whether the input data
contains count data, or TPM abundance. The input data is assumed
to be normalized, and is not adjusted.</p>
<p>Statistical results can be summarized at the gene level, after
applying thresholds for statistical hits in the form of
required adjusted P-value and/or fold change. It may be helpful
to review results per gene, alongside the specific transcript
isoforms which are called statistical hits.</p>
<p>This function may also test the subset of "detected transcripts",
for example based upon the output of <code><a href='defineDetectedTx.html'>defineDetectedTx()</a></code>.
In practice, this strategy removes a number of potential hits
which are otherwise not considered to be "detected". The intent is to
filter out potential hits based upon a noise threshold, which
may be below the practical limit of follow-up validation techniques.
Another goal is the removal of a large number of undetected
transcripts which are effectively not tested, prior to
multiple testing correction.</p>
<p>The input data <code>iMatrixTx</code> should contain rows of data where the
<code><a href='https://www.rdocumentation.org/packages/base/topics/colnames'>rownames(iMatrixTx)</a></code> are represented in <code>tx2geneDF[,txColname]</code>
in order to link rows to genes, using <code>tx2geneDF[,geneColname]</code>.
The rows  of <code>iMatrixTx</code> are expected to contain expression values
per transcript isoform, but may contain alternative measurements
such as: junction expression per gene; exon expression per gene.</p>
<p>Details:</p>
<p>This function provides a wrapper of a basic workflow for differential
isoform analysis, including these basic steps in order:</p>
<dl class='dl-horizontal'>
<dt><code>limma:voom()</code></dt><dd><p>(Optional.) This step is enabled with
the argument <code>useVoom=TRUE</code> and should only be used when <code>iMatrixTx</code>
contains count or pseudocount data. When using TPM or FPKM values,
set <code>useVoom=FALSE</code>.</p></dd>
<dt><code><a href='https://www.rdocumentation.org/packages/limma/topics/lmFit'>limma::lmFit()</a></code></dt><dd><p></p></dd>
<dt><code><a href='https://www.rdocumentation.org/packages/limma/topics/contrasts.fit'>limma::contrasts.fit()</a></code></dt><dd><p></p></dd>
<dt><code><a href='https://www.rdocumentation.org/packages/limma/topics/diffSplice'>limma::diffSplice()</a></code></dt><dd><p></p></dd>
<dt><code><a href='https://www.rdocumentation.org/packages/limma/topics/topSplice'>limma::topSplice()</a></code></dt><dd><p>This function is called on each contrast
in order to return a list of data.frames.</p></dd>
</dl>
    <p>Note that <code>iMatrixTx</code> is exponentiated prior to running <code><a href='https://www.rdocumentation.org/packages/limma/topics/voom'>limma::voom()</a></code>
when <code>useVoom=TRUE</code>, for the purpose of calculating a weights matrix.
The original <code>iMatrixTx</code> data is used in <code>limma::lmFits()</code> as-is,
alongside the voom weights. The voom-normalized data is not
used. Therefore, the input data is assumed to be normalized.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>Other jam RNA-seq functions: <code><a href='ale2violin.html'>ale2violin</a></code>,
  <code><a href='codonUsage2df.html'>codonUsage2df</a></code>,
  <code><a href='defineDetectedTx.html'>defineDetectedTx</a></code>,
  <code><a href='detectedTxInfo.html'>detectedTxInfo</a></code>, <code><a href='dna2codon.html'>dna2codon</a></code>,
  <code><a href='groups2contrasts.html'>groups2contrasts</a></code>, <code><a href='jamCai.html'>jamCai</a></code>,
  <code><a href='makeTx2geneFromGtf.html'>makeTx2geneFromGtf</a></code>,
  <code><a href='sortSamples.html'>sortSamples</a></code>, <code><a href='tx2ale.html'>tx2ale</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># example for defining iDesign</span>
<span class='co'># first define a vector of sample groups</span>
<span class='no'>iGroups</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>jamba</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/jamba/topics/nameVector'>nameVector</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/paste'>paste</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/rep'>rep</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"WT"</span>, <span class='st'>"KO"</span>), <span class='kw'>each</span><span class='kw'>=</span><span class='fl'>6</span>),
   <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/rep'>rep</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"Control"</span>, <span class='st'>"Treated"</span>), <span class='kw'>each</span><span class='kw'>=</span><span class='fl'>3</span>),
   <span class='kw'>sep</span><span class='kw'>=</span><span class='st'>"_"</span>));
<span class='no'>iGroups</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/factor'>factor</a></span>(<span class='no'>iGroups</span>, <span class='kw'>levels</span><span class='kw'>=</span><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/unique'>unique</a></span>(<span class='no'>iGroups</span>));
<span class='no'>iGroups</span>;</div><div class='output co'>#&gt; WT_Control_v1 WT_Control_v2 WT_Control_v3 WT_Treated_v1 WT_Treated_v2 
#&gt;    WT_Control    WT_Control    WT_Control    WT_Treated    WT_Treated 
#&gt; WT_Treated_v3 KO_Control_v1 KO_Control_v2 KO_Control_v3 KO_Treated_v1 
#&gt;    WT_Treated    KO_Control    KO_Control    KO_Control    KO_Treated 
#&gt; KO_Treated_v2 KO_Treated_v3 
#&gt;    KO_Treated    KO_Treated 
#&gt; Levels: WT_Control WT_Treated KO_Control KO_Treated</div><div class='input'><span class='co'># next define sample identifiers</span>
<span class='no'>iSamples</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/names'>names</a></span>(<span class='no'>iGroups</span>);

<span class='co'># given a vector of groups, make iDesign</span>
<span class='no'>iDesign</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>stats</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/stats/topics/model.matrix'>model.matrix</a></span>(~<span class='fl'>0</span>+<span class='no'>iGroups</span>);

<span class='co'># It is good practice to rename colnames(iDesign) and rownames(iDesign),</span>
<span class='co'># that is, for the love of all that is good, use colnames and rownames</span>
<span class='co'># that help confirm that these matrices are consistent.</span>
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/colnames'>colnames</a></span>(<span class='no'>iDesign</span>) <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/levels'>levels</a></span>(<span class='no'>iGroups</span>);
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/colnames'>rownames</a></span>(<span class='no'>iDesign</span>) <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/names'>names</a></span>(<span class='no'>iGroups</span>);
<span class='no'>iDesign</span>;</div><div class='output co'>#&gt;               WT_Control WT_Treated KO_Control KO_Treated
#&gt; WT_Control_v1          1          0          0          0
#&gt; WT_Control_v2          1          0          0          0
#&gt; WT_Control_v3          1          0          0          0
#&gt; WT_Treated_v1          0          1          0          0
#&gt; WT_Treated_v2          0          1          0          0
#&gt; WT_Treated_v3          0          1          0          0
#&gt; KO_Control_v1          0          0          1          0
#&gt; KO_Control_v2          0          0          1          0
#&gt; KO_Control_v3          0          0          1          0
#&gt; KO_Treated_v1          0          0          0          1
#&gt; KO_Treated_v2          0          0          0          1
#&gt; KO_Treated_v3          0          0          0          1
#&gt; attr(,"assign")
#&gt; [1] 1 1 1 1
#&gt; attr(,"contrasts")
#&gt; attr(,"contrasts")$iGroups
#&gt; [1] "contr.treatment"
#&gt; </div><div class='input'>
<span class='co'># define contrasts</span>
<span class='co'># the example below includes a two-way contrast, which is a test</span>
<span class='co'># of the pairwise fold changes</span>
<span class='no'>iContrasts</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>limma</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/limma/topics/makeContrasts'>makeContrasts</a></span>(<span class='kw'>contrasts</span><span class='kw'>=</span><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(
   <span class='st'>"WT_Treated-WT_Control"</span>, <span class='st'>"KO_Treated-KO_Control"</span>,
   <span class='st'>"(KO_Treated-KO_Control)-(WT_Treated-WT_Control)"</span>),
   <span class='kw'>levels</span><span class='kw'>=</span><span class='no'>iDesign</span>);
<span class='no'>iContrasts</span>;</div><div class='output co'>#&gt;             Contrasts
#&gt; Levels       WT_Treated-WT_Control KO_Treated-KO_Control
#&gt;   WT_Control                    -1                     0
#&gt;   WT_Treated                     1                     0
#&gt;   KO_Control                     0                    -1
#&gt;   KO_Treated                     0                     1
#&gt;             Contrasts
#&gt; Levels       (KO_Treated-KO_Control)-(WT_Treated-WT_Control)
#&gt;   WT_Control                                               1
#&gt;   WT_Treated                                              -1
#&gt;   KO_Control                                              -1
#&gt;   KO_Treated                                               1</div><div class='input'>
<span class='co'># for validation, verify these constraints:</span>
<span class='co'># - all(rownames(iDesign) == iSamples)</span>
<span class='co'># - colnames(iDesign) == the actual group names</span>
<span class='co'># - all(colnames(iDesign) == rownames(iContrasts))</span>
<span class='co'># you can see which samples are included in each test with crossproduct:</span>
<span class='no'>iDesign</span> <span class='kw'>%*%</span> <span class='no'>iContrasts</span>;</div><div class='output co'>#&gt;                Contrasts
#&gt;                 WT_Treated-WT_Control KO_Treated-KO_Control
#&gt;   WT_Control_v1                    -1                     0
#&gt;   WT_Control_v2                    -1                     0
#&gt;   WT_Control_v3                    -1                     0
#&gt;   WT_Treated_v1                     1                     0
#&gt;   WT_Treated_v2                     1                     0
#&gt;   WT_Treated_v3                     1                     0
#&gt;   KO_Control_v1                     0                    -1
#&gt;   KO_Control_v2                     0                    -1
#&gt;   KO_Control_v3                     0                    -1
#&gt;   KO_Treated_v1                     0                     1
#&gt;   KO_Treated_v2                     0                     1
#&gt;   KO_Treated_v3                     0                     1
#&gt;                Contrasts
#&gt;                 (KO_Treated-KO_Control)-(WT_Treated-WT_Control)
#&gt;   WT_Control_v1                                               1
#&gt;   WT_Control_v2                                               1
#&gt;   WT_Control_v3                                               1
#&gt;   WT_Treated_v1                                              -1
#&gt;   WT_Treated_v2                                              -1
#&gt;   WT_Treated_v3                                              -1
#&gt;   KO_Control_v1                                              -1
#&gt;   KO_Control_v2                                              -1
#&gt;   KO_Control_v3                                              -1
#&gt;   KO_Treated_v1                                               1
#&gt;   KO_Treated_v2                                               1
#&gt;   KO_Treated_v3                                               1</div><div class='input'>
<span class='co'>## Another efficient way to define iDesign and iContrasts:</span>
<span class='no'>iDC</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>splicejam</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/splicejam/topics/groups2contrasts'>groups2contrasts</a></span>(<span class='no'>iGroups</span>, <span class='kw'>returnDesign</span><span class='kw'>=</span><span class='fl'>TRUE</span>);
<span class='no'>iDesign</span> <span class='kw'>&lt;-</span> <span class='no'>iDC</span>$<span class='no'>iDesign</span>;
<span class='no'>iContrasts</span> <span class='kw'>&lt;-</span> <span class='no'>iDC</span>$<span class='no'>iContrasts</span>;</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by James M. Ward.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

